
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Bayesian Model Comparison &#8212; Statistical backgrounds for genetics trainees</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Bayesian_model_comparison';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bayesian Mixture Model" href="Bayesian_mixture_model.html" />
    <link rel="prev" title="Bayesian Multivariate Normal Mean Model" href="Bayesian_multivariate_normal_mean_model.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="README.html">
  
  
  
  
  
  
    <p class="title logo__title">Statistical backgrounds for genetics trainees</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Popular Topics and Related Notebooks</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="notations.html">Notations</a></li>
<li class="toctree-l1"><a class="reference internal" href="popular_topics.html">Popular Topics in Statistical Genetics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Genetics Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genotype_coding.html">Genotype Coding</a></li>







<li class="toctree-l1"><a class="reference internal" href="minor_allele_frequency.html">Minor Allele Frequency</a></li>





<li class="toctree-l1"><a class="reference internal" href="Hardy_Weinberg_equilibrium.html">Hardy-Weinberg Equilibrium</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Variants</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium.html">Linkage Disequilibrium</a></li>






<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium_score.html">Linkage Disequilibrium Score</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Individuals</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_relationship_matrix.html">Genetic Relationship Matrix</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Fixed Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ordinary_least_squares.html">Ordinary Least Squares</a></li>






<li class="toctree-l1"><a class="reference internal" href="odds_ratio.html">Odds Ratio</a></li>






<li class="toctree-l1"><a class="reference internal" href="marginal_joint_effects.html">Marginal and Joint Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="summary_statistics.html">Summary Statistics</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Random Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="random_effects.html">Random Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="proportion_of_variance_explained.html">Proportion of Variance Explained and Heritability</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Linear Mixed Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linear_mixed_model.html">Linear Mixed Model</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Covariates</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="confounder.html">Confounder</a></li>





<li class="toctree-l1"><a class="reference internal" href="collider.html">Collider</a></li>






<li class="toctree-l1"><a class="reference internal" href="mediator.html">Mediator</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Meta-analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="meta_analysis_fixed_effect.html">Meta-Analysis Fixed Effect</a></li>





<li class="toctree-l1"><a class="reference internal" href="meta_analysis_random_effect.html">Meta-Analysis Random Effect</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Likelihood</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="likelihood.html">Likelihood</a></li>






<li class="toctree-l1"><a class="reference internal" href="maximum_likelihood_estimation.html">Maximum Likelihood Estimation</a></li>





<li class="toctree-l1"><a class="reference internal" href="likelihood_ratio.html">Likelihood Ratio</a></li>







<li class="toctree-l1"><a class="reference internal" href="expectation_maximum.html">Expectation-Maximum Algorithm</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Versus Frequentist</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_versus_frequentist.html">Bayesian versus Frequentist</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bayes_rule.html">Bayes Rule</a></li>





<li class="toctree-l1"><a class="reference internal" href="Bayes_factor.html">Bayes Factor</a></li>





<li class="toctree-l1"><a class="reference internal" href="p_value_and_Bayesian_hypothesis_testing.html"><span class="math notranslate nohighlight">\(p\)</span>-value and Bayesian Hypothesis Testing</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_normal_mean_model.html">Bayesian Normal Mean Model</a></li>





<li class="toctree-l1"><a class="reference internal" href="Bayesian_multivariate_normal_mean_model.html">Bayesian Multivariate Normal Mean Model</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiple Bayesian Models</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Bayesian Model Comparison</a></li>





<li class="toctree-l1"><a class="reference internal" href="Bayesian_mixture_model.html">Bayesian Mixture Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Latent Structures in Data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="factor_analysis.html">Factor Analysis</a></li>





<li class="toctree-l1"><a class="reference internal" href="principal_component_analysis.html">Principal Component Analysis</a></li>







<li class="toctree-l1"><a class="reference internal" href="hidden_Markov_model.html">Hidden Markov Model</a></li>






</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Bayesian_model_comparison.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Bayesian Model Comparison</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Bayesian Model Comparison</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bayes-factor">Bayes Factor</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-odds">Posterior Odds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-the-marginal-likelihood">Computing the Marginal Likelihood</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-fine-mapping">Example 1 – Fine Mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#prior-distribution">Prior Distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihoods-and-bayes-factors">Computing Marginal Likelihoods and Bayes Factors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-distribution">Posterior Distribution</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-detecting-pleiotropy">Example 2 – Detecting pleiotropy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Prior Distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihood-and-bayes-factor">Computing Marginal Likelihood and Bayes Factor</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Posterior Distribution</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="bayesian-model-comparison">
<h1>Bayesian Model Comparison<a class="headerlink" href="#bayesian-model-comparison" title="Link to this heading">#</a></h1>
<p>Bayesian model comparison quantifies uncertainty about which model generated our data by computing marginal likelihoods that measure how well each model predicts the observations after integrating over all possible parameter values.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="graphical-summary">
<h1>Graphical Summary<a class="headerlink" href="#graphical-summary" title="Link to this heading">#</a></h1>
<p><img alt="Fig" src="_images/Slide28.png" /></p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="key-formula">
<h1>Key Formula<a class="headerlink" href="#key-formula" title="Link to this heading">#</a></h1>
<p>When we cannot condition on which model (or model family) <span class="math notranslate nohighlight">\(\text{M}_1\)</span> or <span class="math notranslate nohighlight">\(\text{M}_2\)</span> generated our data, we quantify our uncertainty about model choice by calculating <strong>posterior odds</strong>. These represent our updated uncertainty about competing models after observing the data:</p>
<div class="math notranslate nohighlight">
\[
\frac{P(\text{M}_1|D)}{P(\text{M}_2|D)} = \frac{P(\text{D}|\text{M}_1)P(\text{M}_1)}{P(\text{D}|\text{M}_2)P(\text{M}_2)}
\]</div>
<p>This can be decomposed as:</p>
<div class="math notranslate nohighlight">
\[
\text{Posterior Odds} = \text{Bayes Factor} \times \text{Prior Odds}
\]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(P(\text{M}_1|D)\)</span> and <span class="math notranslate nohighlight">\(P(\text{M}_2|D)\)</span> are our <strong>posterior uncertainties</strong> about models <span class="math notranslate nohighlight">\(\text{M}_1\)</span> and <span class="math notranslate nohighlight">\(\text{M}_2\)</span> after conditioning on the observed data</p></li>
<li><p><span class="math notranslate nohighlight">\(\frac{P(\text{D}|\text{M}_1)}{P(\text{D}|\text{M}_2)}\)</span> is the <strong>Bayes factor</strong>, which quantifies the relative evidence the data provides for each model</p></li>
<li><p><span class="math notranslate nohighlight">\(P(\text{M}_1)\)</span> and <span class="math notranslate nohighlight">\(P(\text{M}_2)\)</span> are our <strong>prior uncertainties</strong> about each model before conditioning on any data, representing what we cannot know a priori about model choice</p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="technical-details">
<h1>Technical Details<a class="headerlink" href="#technical-details" title="Link to this heading">#</a></h1>
<section id="bayes-factor">
<h2>Bayes Factor<a class="headerlink" href="#bayes-factor" title="Link to this heading">#</a></h2>
<p>Recall in <a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_factor.html">Lecture: Bayes factor</a>, we learn that the <strong>Bayes factor</strong> quantifies the relative evidence data provides for competing models:</p>
<div class="math notranslate nohighlight">
\[
\text{BF}_{21} = \frac{L(\text{M}_2|\text{D})}{L(\text{M}_1|\text{D})} ={\frac {\int \Pr(\theta_{2}|M_{2})\Pr(D|\theta_{2},M_{2})\,d\theta _{2}}{\int \Pr(\theta_{1}|M_{1})\Pr(D|\theta_{1},M_{1})\,d\theta _{1}}}
\]</div>
<p><strong>Interpretation</strong>:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} &gt; 1\)</span>: Data favors <span class="math notranslate nohighlight">\(\text{M}_2\)</span> over <span class="math notranslate nohighlight">\(\text{M}_1\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} &lt; 1\)</span>: Data favors <span class="math notranslate nohighlight">\(\text{M}_1\)</span> over <span class="math notranslate nohighlight">\(\text{M}_2\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} = 1\)</span>: Data uninformative for model choice</p></li>
</ul>
</section>
<section id="posterior-odds">
<h2>Posterior Odds<a class="headerlink" href="#posterior-odds" title="Link to this heading">#</a></h2>
<p>The Bayes factor updates prior beliefs about models:</p>
<div class="math notranslate nohighlight">
\[
\frac{P(\text{M}_2|D)}{P(\text{M}_1|D)} = \text{BF}_{21} \times \frac{P(\text{M}_2)}{P(\text{M}_1)}
\]</div>
<p>This shows how conditioning on data transforms our uncertainty:</p>
<div class="math notranslate nohighlight">
\[
\text{Posterior Uncertainty} = \text{Evidence Update} \times \text{Prior Uncertainty}
\]</div>
</section>
<section id="computing-the-marginal-likelihood">
<h2>Computing the Marginal Likelihood<a class="headerlink" href="#computing-the-marginal-likelihood" title="Link to this heading">#</a></h2>
<p>The integral <span class="math notranslate nohighlight">\(\int \Pr(\theta|\text{M})\Pr(\text{D}|\theta,M)\,d\theta\)</span> is the <strong>marginal likelihood</strong> – the probability of data under a model, averaged over all parameter values.</p>
<p><strong>Analytical Solutions</strong></p>
<p>With conjugate priors (e.g., normal-normal), we obtain closed-form marginal likelihoods.</p>
<p><strong>Approximation Methods</strong></p>
<p>For non-conjugate cases, we need numerical approximations:</p>
<ul class="simple">
<li><p><strong>Laplace approximation</strong>: Assumes concentrated posterior around the mode</p></li>
<li><p><strong>MCMC methods</strong>: Chib’s method, thermodynamic integration</p></li>
<li><p><strong>Nested sampling</strong>: Systematically explores the likelihood surface</p></li>
</ul>
<p>The choice reflects a trade-off between computational feasibility and how thoroughly we can characterize the parameter uncertainty in our model comparison.</p>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="related-topics">
<h1>Related Topics<a class="headerlink" href="#related-topics" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/ordinary_least_squares.html">ordinary least squares</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/odds_ratio.html">odds ratio</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/summary_statistics.html">summary statistics</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/linkage_disequilibrium.html">linkage disequilibrium</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/marginal_joint_effects.html">marginal and joint effects</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_factor.html">Bayes factor</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_rule.html">Bayes rule</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_normal_mean_model.html">Bayesian normal mean model</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_multivariate_normal_mean_model.html">Bayesian multivariate normal mean model</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_mixture_model.html">Bayesian mixture model</a></p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="example">
<h1>Example<a class="headerlink" href="#example" title="Link to this heading">#</a></h1>
<section id="example-1-fine-mapping">
<h2>Example 1 – Fine Mapping<a class="headerlink" href="#example-1-fine-mapping" title="Link to this heading">#</a></h2>
<p>Recall we studied how LD affects our understanding of genetic effects in Example 1 in the <a class="reference external" href="https://statfungen.github.io/statgen-primer/marginal_joint_effects.html">Lecture: marginal and joint effects</a>. We explored two approaches: fitting a joint model with all variants, and performing conditional analysis by iteratively conditioning on the variant with the smallest p-value. Both treated effects as fixed parameters.</p>
<p>In genomics with hundreds of correlated variants in LD, these approaches have critical limitations. Joint models are often infeasible – there can easily be more variants than samples, making the model unidentifiable. Even when feasible, these models become unstable due to overfitting and collinearity. Conditional analysis, while practical, commits to a single “best” variant at each step without quantifying uncertainty about whether that variant is truly causal.</p>
<p>This is the <strong>fine-mapping</strong> problem: variants in LD are inherited together, creating statistical association between all linked variants even when only one is truly causal. We face fundamental uncertainty about which variant drives the association.</p>
<p>Bayesian model comparison provides a principled framework for this setting. Rather than making sequential conditioning decisions or attempting unstable joint models, we treat causal variant identity as uncertain and systematically evaluate the posterior probability that each variant is causal, properly accounting for the correlation structure in the data.</p>
<p>We define our competing models:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\text{M}_0\)</span>: Null model – none of the variants are causal</p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_1\)</span>: Variant 1 is the causal variant</p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_2\)</span>: Variant 2 is the causal variant</p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_3\)</span>: Variant 3 is the causal variant</p></li>
</ul>
<section id="setup">
<h3>Setup<a class="headerlink" href="#setup" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">9</span><span class="p">)</span><span class="w"> </span>

<span class="c1"># Define genotypes for 20 individuals at 3 variants</span>
<span class="c1"># Create correlated genotypes to simulate linkage disequilibrium</span>
<span class="n">N</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span>
<span class="n">M</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span>

<span class="c1"># Generate correlated genotype data</span>
<span class="c1"># Variant 1 is the true causal variant</span>
<span class="c1"># Variants 2 and 3 are in LD with variant 1</span>
<span class="n">variant1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">))</span>
<span class="c1"># Create LD: variants 2 and 3 are correlated with variant 1</span>
<span class="n">variant2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">runif</span><span class="p">(</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="n">variant3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">runif</span><span class="p">(</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.8</span><span class="p">,</span><span class="w"> </span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>

<span class="n">Xraw_additive</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="n">variant2</span><span class="p">,</span><span class="w"> </span><span class="n">variant3</span><span class="p">)</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Individual&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">N</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Variant&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">M</span><span class="p">)</span>

<span class="c1"># Standardize genotypes</span>
<span class="n">X</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">,</span><span class="w"> </span><span class="n">center</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Generate phenotype where only Variant 1 has a true causal effect</span>
<span class="n">true_beta1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1.5</span>
<span class="n">epsilon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rnorm</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span>
<span class="n">Y_raw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">X</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">epsilon</span><span class="w">  </span><span class="c1"># Only variant 1 affects the trait</span>

<span class="c1"># Standardize phenotype</span>
<span class="n">Y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">Y_raw</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Notice how we’ve created linkage disequilibrium - variants 2 and 3 are correlated with variant 1, but only variant 1 actually causes the phenotype.</p>
<p>Now we need to define our statistical models. For each model <span class="math notranslate nohighlight">\(M_k\)</span> (where <span class="math notranslate nohighlight">\(k\)</span> indexes the causal variant), we assume:</p>
<ul class="simple">
<li><p><strong>Likelihood</strong>: <span class="math notranslate nohighlight">\(y_i | \beta, \sigma^2, M_k \sim N(X_k\beta, \sigma^2)\)</span>, where <span class="math notranslate nohighlight">\(X_k\)</span> is the genotype for variant <span class="math notranslate nohighlight">\(k\)</span></p></li>
<li><p><strong>Prior for genetic effect</strong>: <span class="math notranslate nohighlight">\(\beta | M_k \sim N(0, \tau^2)\)</span> with <span class="math notranslate nohighlight">\(\tau^2 = 1\)</span></p></li>
<li><p><strong>Prior for error variance</strong>: <span class="math notranslate nohighlight">\(\sigma^2 | M_k \sim \text{Inverse-Gamma}(a, b)\)</span> with <span class="math notranslate nohighlight">\(a = 2\)</span>, <span class="math notranslate nohighlight">\(b = 1\)</span></p></li>
</ul>
</section>
<section id="prior-distribution">
<h3>Prior Distribution<a class="headerlink" href="#prior-distribution" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Bayesian model setup</span>
<span class="c1"># We&#39;ll use conjugate priors for computational convenience</span>

<span class="c1"># Prior parameters for genetic effect (beta)</span>
<span class="n">beta_prior_mean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="n">beta_prior_var</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w">  </span><span class="c1"># tau^2 in the notes</span>

<span class="c1"># Assume equal prior probabilities for all models</span>
<span class="n">prior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0.25</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">  </span><span class="c1"># Equal priors for M0, M1, M2, M3</span>

<span class="c1"># Prior parameters for error variance (sigma^2)</span>
<span class="c1"># Using inverse-gamma: IG(a, b)</span>
<span class="n">sigma2_prior_a</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span>
<span class="n">sigma2_prior_b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="computing-marginal-likelihoods-and-bayes-factors">
<h3>Computing Marginal Likelihoods and Bayes Factors<a class="headerlink" href="#computing-marginal-likelihoods-and-bayes-factors" title="Link to this heading">#</a></h3>
<p>The key to Bayesian model comparison is computing the marginal likelihood for each model. This quantity represents how well each model explains the data after integrating out uncertainty in the parameters.</p>
<p>For our conjugate normal-inverse-gamma setup, the marginal likelihood has a closed-form solution, allowing us to compute it analytically rather than relying on numerical approximations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to compute marginal likelihood for each single-variant model</span>
<span class="c1"># Uses conjugate normal-inverse-gamma priors for analytical tractability</span>
<span class="n">compute_marginal_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">X_variant</span><span class="p">,</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Posterior parameters under normal-inverse-gamma conjugacy</span>
<span class="w">  </span><span class="n">V_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">X_variant</span><span class="o">^</span><span class="m">2</span><span class="p">))</span>
<span class="w">  </span><span class="n">mu_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">V_n</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="n">beta_prior_mean</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">X_variant</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">Y</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">a_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sigma2_prior_a</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n</span><span class="o">/</span><span class="m">2</span>
<span class="w">  </span><span class="n">b_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sigma2_prior_b</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">Y</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">                                  </span><span class="n">beta_prior_mean</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">                                  </span><span class="n">mu_n</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">V_n</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Log marginal likelihood (closed-form for conjugate priors)</span>
<span class="w">  </span><span class="n">log_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="o">-</span><span class="n">n</span><span class="o">/</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="kc">pi</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">            </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">V_n</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">            </span><span class="n">sigma2_prior_a</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">sigma2_prior_b</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">            </span><span class="n">a_n</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">b_n</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">            </span><span class="nf">lgamma</span><span class="p">(</span><span class="n">a_n</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nf">lgamma</span><span class="p">(</span><span class="n">sigma2_prior_a</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">log_ml</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Compute marginal likelihoods for each model</span>
<span class="n">log_ml_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">),</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">  </span><span class="c1"># Null model (no causal variant)</span>
<span class="n">log_ml_M1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">      </span><span class="c1"># Variant 1 is causal</span>
<span class="n">log_ml_M2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">      </span><span class="c1"># Variant 2 is causal  </span>
<span class="n">log_ml_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">3</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">      </span><span class="c1"># Variant 3 is causal</span>

<span class="c1"># Store results</span>
<span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">log_ml_M0</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M1</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;M0 (Null)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M1 (Var1)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 (Var2)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M3 (Var3)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can compute Bayes factors to compare models.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute all Bayes factors</span>
<span class="n">BF_M1_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>
<span class="n">BF_M2_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>
<span class="n">BF_M3_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M3</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>
<span class="n">BF_M1_vs_M2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="p">)</span>
<span class="n">BF_M1_vs_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>
<span class="n">BF_M2_vs_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>

<span class="c1"># Create single dataframe with all comparisons</span>
<span class="n">bf_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Comparison</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;M1 vs M0&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 vs M0&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M3 vs M0&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                 </span><span class="s">&quot;M1 vs M2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M1 vs M3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 vs M3&quot;</span><span class="p">),</span>
<span class="w">  </span><span class="n">Bayes_Factor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">BF_M1_vs_M0</span><span class="p">,</span><span class="w"> </span><span class="n">BF_M2_vs_M0</span><span class="p">,</span><span class="w"> </span><span class="n">BF_M3_vs_M0</span><span class="p">,</span>
<span class="w">                   </span><span class="n">BF_M1_vs_M2</span><span class="p">,</span><span class="w"> </span><span class="n">BF_M1_vs_M3</span><span class="p">,</span><span class="w"> </span><span class="n">BF_M2_vs_M3</span><span class="p">),</span>
<span class="w">  </span><span class="n">Log_BF</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">,</span>
<span class="w">             </span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>
<span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Bayes Factors (BF &gt; 1 favors first model):&quot;</span><span class="p">)</span>
<span class="n">bf_results</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Bayes Factors (BF &gt; 1 favors first model):&quot;
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 3</caption>
<thead>
	<tr><th scope=col>Comparison</th><th scope=col>Bayes_Factor</th><th scope=col>Log_BF</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>M1 vs M0</td><td>6.783402e+06</td><td>15.729989</td></tr>
	<tr><td>M2 vs M0</td><td>4.059365e+00</td><td> 1.401026</td></tr>
	<tr><td>M3 vs M0</td><td>7.763487e+03</td><td> 8.957187</td></tr>
	<tr><td>M1 vs M2</td><td>1.671050e+06</td><td>14.328963</td></tr>
	<tr><td>M1 vs M3</td><td>8.737571e+02</td><td> 6.772802</td></tr>
	<tr><td>M2 vs M3</td><td>5.228790e-04</td><td>-7.556160</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="posterior-distribution">
<h3>Posterior Distribution<a class="headerlink" href="#posterior-distribution" title="Link to this heading">#</a></h3>
<p>We now combine the prior probabilities with the Bayes factors to obtain posterior probabilities for each model. This tells us, given the data, how probable each model is.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute posterior probabilities</span>
<span class="c1"># Use log-sum-exp trick to avoid numerical overflow</span>
<span class="n">max_log_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">)</span>
<span class="n">scaled_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log_ml</span><span class="p">)</span>

<span class="c1"># Apply Bayes&#39; rule: posterior ∝ prior × likelihood</span>
<span class="n">unnormalized_posterior</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prior_probs</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">scaled_ml</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unnormalized_posterior</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">unnormalized_posterior</span><span class="p">)</span>

<span class="c1"># Create results dataframe</span>
<span class="n">posterior_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;M0 (Null)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M1 (Var1)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 (Var2)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M3 (Var3)&quot;</span><span class="p">),</span>
<span class="w">  </span><span class="n">Prior</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prior_probs</span><span class="p">,</span>
<span class="w">  </span><span class="n">Log_Marginal_Likelihood</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="p">,</span>
<span class="w">  </span><span class="n">Posterior</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">posterior_probs</span>
<span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Posterior Model Probabilities:&quot;</span><span class="p">)</span>
<span class="n">posterior_results</span>

<span class="c1"># Identify most probable model</span>
<span class="n">best_model_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which.max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Most probable model:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">posterior_results</span><span class="o">$</span><span class="n">Model</span><span class="p">[</span><span class="n">best_model_idx</span><span class="p">]))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Posterior probability:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">[</span><span class="n">best_model_idx</span><span class="p">],</span><span class="w"> </span><span class="m">4</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Posterior Model Probabilities:&quot;
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 4 × 4</caption>
<thead>
	<tr><th></th><th scope=col>Model</th><th scope=col>Prior</th><th scope=col>Log_Marginal_Likelihood</th><th scope=col>Posterior</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>M0 (Null)</th><td>M0 (Null)</td><td>0.25</td><td>-29.09297</td><td>1.472500e-07</td></tr>
	<tr><th scope=row>M1 (Var1)</th><td>M1 (Var1)</td><td>0.25</td><td>-13.36298</td><td>9.988561e-01</td></tr>
	<tr><th scope=row>M2 (Var2)</th><td>M2 (Var2)</td><td>0.25</td><td>-27.69194</td><td>5.977415e-07</td></tr>
	<tr><th scope=row>M3 (Var3)</th><td>M3 (Var3)</td><td>0.25</td><td>-20.13578</td><td>1.143174e-03</td></tr>
</tbody>
</table>
</div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Most probable model: M1 (Var1)&quot;
[1] &quot;Posterior probability: 0.9989&quot;
</pre></div>
</div>
</div>
</div>
<p>In this example, we demonstrated how Bayesian model comparison solves the fine-mapping problem. Even when variants are in LD and all show association with the phenotype, our method identified the true causal variant by systematically comparing models and quantifying uncertainty in parameter estimates.</p>
<p>The key insight is that the Bayes factor naturally penalizes model complexity – a model must fit the data substantially better to justify additional parameters. This helps distinguish true causation from correlation due to LD.</p>
<p>This approach forms the foundation for modern fine-mapping methods in genomics.</p>
</section>
</section>
<section id="example-2-detecting-pleiotropy">
<h2>Example 2 – Detecting pleiotropy<a class="headerlink" href="#example-2-detecting-pleiotropy" title="Link to this heading">#</a></h2>
<p>Recall our <a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_multivariate_normal_mean_model.html">Lecture: Bayesian multivariate normal mean model</a> example where one genetic variant could affect two traits (height and weight).</p>
<p>In practice, this variant could affect height and weight <span class="math notranslate nohighlight">\(\boldsymbol{\beta} = (\beta_1, \beta_2)\)</span> in multiple ways:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\text{M}_0\)</span> (No effect): <span class="math notranslate nohighlight">\(\boldsymbol{\beta} = (0, 0)\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_1\)</span> (Height Only): <span class="math notranslate nohighlight">\(\beta_2 = 0\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_2\)</span> (Weight Only): <span class="math notranslate nohighlight">\(\beta_1 = 0\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_3\)</span> (Perfect Correlation): <span class="math notranslate nohighlight">\(\text{cor}(\beta_1, \beta_2) = 1\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_4\)</span> (Weak Correlation): <span class="math notranslate nohighlight">\(\text{cor}(\beta_1, \beta_2) = 0.1\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_5\)</span> (Medium Correlation): <span class="math notranslate nohighlight">\(\text{cor}(\beta_1, \beta_2) = 0.5\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{M}_6\)</span> (Strong Correlation): <span class="math notranslate nohighlight">\(\text{cor}(\beta_1, \beta_2) = 0.8\)</span></p></li>
<li><p>…</p></li>
</ul>
<p>We cannot condition on which model generated our data. Bayesian model comparison allows us to quantify uncertainty across these competing hypotheses by computing marginal likelihoods <span class="math notranslate nohighlight">\(P(\text{D}|\text{M})\)</span> for each model, integrating over all possible parameter values.</p>
<section id="id1">
<h3>Setup<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<p>Let’s simulate data where we know the truth - the variant has a medium correlation effect (M5).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">library</span><span class="p">(</span><span class="n">MASS</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">mvtnorm</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">91</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Data Generation for Bayesian Model Comparison</span>
<span class="c1"># Scenario: One variant affecting two traits (height and weight) with correlated effects</span>

<span class="c1"># ============================================================================</span>
<span class="c1"># Parameters</span>
<span class="c1"># ============================================================================</span>
<span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span><span class="w">                  </span><span class="c1"># Sample size</span>
<span class="n">true_correlation</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.5</span><span class="w">   </span><span class="c1"># True correlation between effects on height and weight (Model 5)</span>
<span class="n">effect_variance</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.25</span><span class="w">   </span><span class="c1"># Variance for each effect</span>

<span class="c1"># ============================================================================</span>
<span class="c1"># Step 1: Generate correlated effect sizes</span>
<span class="c1"># ============================================================================</span>
<span class="c1"># Covariance matrix for bivariate effects</span>
<span class="n">Sigma_effects</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span>
<span class="w">                         </span><span class="n">true_correlation</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                         </span><span class="n">true_correlation</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span>
<span class="w">                         </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">                       </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="c1"># Draw true effects from multivariate normal</span>
<span class="n">beta_true</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mvrnorm</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">mu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="n">Sigma</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sigma_effects</span><span class="p">)</span>
<span class="n">true_beta_height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">beta_true</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">  </span><span class="c1"># Effect on height</span>
<span class="n">true_beta_weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">beta_true</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w">  </span><span class="c1"># Effect on weight</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Generated effect sizes:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Effect on height (beta1):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">true_beta_height</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Effect on weight (beta2):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">true_beta_weight</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>

<span class="c1"># ============================================================================</span>
<span class="c1"># Step 2: Generate genotype data</span>
<span class="c1"># ============================================================================</span>
<span class="c1"># Genotype coding: 0, 1, 2 copies of effect allele</span>
<span class="n">minor_allele_freq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.3</span>
<span class="n">genotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbinom</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">minor_allele_freq</span><span class="p">)</span>

<span class="c1"># ============================================================================</span>
<span class="c1"># Step 3: Generate phenotype data</span>
<span class="c1"># ============================================================================</span>
<span class="c1"># Residual covariance matrix (environmental effects)</span>
<span class="n">Sigma_residual</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1.0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">,</span><span class="w">    </span><span class="c1"># Height variance = 1.0</span>
<span class="w">                          </span><span class="m">0.2</span><span class="p">,</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w">    </span><span class="c1"># Weight variance = 0.8, covariance = 0.2</span>
<span class="w">                        </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="c1"># Generate correlated residual errors</span>
<span class="n">residual_errors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mvrnorm</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">mu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">Sigma</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sigma_residual</span><span class="p">)</span>

<span class="c1"># Generate phenotypes: Y = baseline + genetic effect + residual error</span>
<span class="n">baseline_height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">170</span><span class="w">  </span><span class="c1"># cm</span>
<span class="n">baseline_weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">70</span><span class="w">   </span><span class="c1"># kg</span>

<span class="n">height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">baseline_height</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">genotypes</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta_height</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">residual_errors</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span>
<span class="n">weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">baseline_weight</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">genotypes</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta_weight</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">residual_errors</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span>

<span class="c1"># Combine and center phenotypes</span>
<span class="n">phenotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">height</span><span class="p">,</span><span class="w"> </span><span class="n">weight</span><span class="p">)</span>
<span class="n">phenotypes_scaled</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">phenotypes_scaled</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Height&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Weight&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated effect sizes:
Effect on height (beta1): 2.181 
Effect on weight (beta2): 4.416 
</pre></div>
</div>
</div>
</div>
</section>
<section id="id2">
<h3>Prior Distribution<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>We’ll define all possible models as prior covariance matrices for the effect sizes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># ============================================================================</span>
<span class="c1"># Prior Setup: Define Models and Prior Probabilities</span>
<span class="c1"># ============================================================================</span>

<span class="c1"># Define prior covariance matrices for beta = (beta_1, beta_2)</span>
<span class="c1"># Each model specifies a different structure for how effects covary</span>
<span class="n">models</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span>
<span class="w">  </span><span class="n">M0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span>
<span class="w">                </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Null: no effect</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span>
<span class="w">                </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Height only: beta_2 = 0</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span>
<span class="w">                </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Weight only: beta_1 = 0</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M3</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Perfect correlation: beta_2 = beta_1</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M4</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                </span><span class="m">0.1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Weak correlation: cor = 0.1</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M5</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">  </span><span class="c1"># Medium correlation: cor = 0.5</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">M6</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.8</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                </span><span class="m">0.8</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span>
<span class="w">              </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">   </span><span class="c1"># Strong correlation: cor = 0.8</span>
<span class="p">)</span>

<span class="c1"># Initialize results dataframe with model names and equal priors</span>
<span class="n">results_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">),</span>
<span class="w">  </span><span class="n">Prior</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="nf">length</span><span class="p">(</span><span class="n">models</span><span class="p">),</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">models</span><span class="p">)),</span>
<span class="w">  </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span>
<span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nPrior probabilities (equal for all models):\n&quot;</span><span class="p">)</span>
<span class="n">results_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Prior probabilities (equal for all models):
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 7 × 2</caption>
<thead>
	<tr><th scope=col>Model</th><th scope=col>Prior</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>M0</td><td>0.1428571</td></tr>
	<tr><td>M1</td><td>0.1428571</td></tr>
	<tr><td>M2</td><td>0.1428571</td></tr>
	<tr><td>M3</td><td>0.1428571</td></tr>
	<tr><td>M4</td><td>0.1428571</td></tr>
	<tr><td>M5</td><td>0.1428571</td></tr>
	<tr><td>M6</td><td>0.1428571</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to compute prior probability when we condition on specific beta values</span>
<span class="n">log_prior</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">component_cov_matrix</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">all</span><span class="p">(</span><span class="n">component_cov_matrix</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># For null model: prior probability is 1 if beta = (0,0), 0 otherwise</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">all</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.01</span><span class="p">),</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="kc">Inf</span><span class="p">))</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Add small regularization for numerical stability</span>
<span class="w">  </span><span class="n">cov_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">component_cov_matrix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">diag</span><span class="p">(</span><span class="m">1e-6</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="w">  </span><span class="n">lp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">-0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="nf">solve</span><span class="p">(</span><span class="n">cov_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">beta</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">        </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">det</span><span class="p">(</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="kc">pi</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">cov_matrix</span><span class="p">))</span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">lp</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="computing-marginal-likelihood-and-bayes-factor">
<h3>Computing Marginal Likelihood and Bayes Factor<a class="headerlink" href="#computing-marginal-likelihood-and-bayes-factor" title="Link to this heading">#</a></h3>
<p>When we cannot condition on the true parameter values, we must integrate over all possibilities to compute marginal likelihoods. This integration captures our uncertainty about parameters within each model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># ============================================================================</span>
<span class="c1"># Computing Marginal Likelihoods</span>
<span class="c1"># ============================================================================</span>

<span class="c1"># Compute marginal likelihood: P(Data | M) = integral of P(Data | beta, M) * P(beta | M)</span>
<span class="n">compute_marginal_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">prior_cov</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">models</span><span class="p">[[</span><span class="n">model_name</span><span class="p">]]</span>

<span class="w">  </span><span class="c1"># Special case: Null model has beta = (0, 0) with probability 1</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">model_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;M0&quot;</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
<span class="w">    </span><span class="n">log_lik</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_log_likelihood</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span>
<span class="w">    </span><span class="n">log_prior_val</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_prior</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">prior_cov</span><span class="p">)</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">log_lik</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">log_prior_val</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>

<span class="w">  </span><span class="c1"># Non-null models: Integrate over all possible beta values</span>
<span class="w">  </span><span class="c1"># Create grid for numerical integration</span>
<span class="w">  </span><span class="n">beta_range</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">-8</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">length.out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">)</span>
<span class="w">  </span><span class="n">grid</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">expand.grid</span><span class="p">(</span><span class="n">beta1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_range</span><span class="p">,</span><span class="w"> </span><span class="n">beta2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_range</span><span class="p">)</span>
<span class="w">  </span><span class="n">grid_spacing</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="n">beta_range</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">beta_range</span><span class="p">[</span><span class="m">1</span><span class="p">])</span><span class="o">^</span><span class="m">2</span>

<span class="w">  </span><span class="c1"># Evaluate integrand at each grid point: likelihood * prior</span>
<span class="w">  </span><span class="n">log_integrand</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">apply</span><span class="p">(</span><span class="n">grid</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">beta_vals</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">beta_vals</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">beta_vals</span><span class="p">[</span><span class="m">2</span><span class="p">])</span>
<span class="w">    </span><span class="n">log_lik</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_log_likelihood</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span>
<span class="w">    </span><span class="n">log_prior_val</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_prior</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">prior_cov</span><span class="p">)</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">log_lik</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">log_prior_val</span><span class="p">)</span>
<span class="w">  </span><span class="p">})</span>

<span class="w">  </span><span class="c1"># Numerical integration using log-sum-exp trick for numerical stability</span>
<span class="w">  </span><span class="n">max_log</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_integrand</span><span class="p">[</span><span class="nf">is.finite</span><span class="p">(</span><span class="n">log_integrand</span><span class="p">)])</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.finite</span><span class="p">(</span><span class="n">max_log</span><span class="p">))</span><span class="w"> </span><span class="kr">return</span><span class="p">(</span><span class="o">-</span><span class="kc">Inf</span><span class="p">)</span>

<span class="w">  </span><span class="n">log_sum</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">max_log</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="nf">exp</span><span class="p">(</span><span class="n">log_integrand</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log</span><span class="p">),</span><span class="w"> </span><span class="n">na.rm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="w">  </span><span class="n">log_marginal</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log_sum</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">grid_spacing</span><span class="p">)</span><span class="w">  </span><span class="c1"># Account for grid spacing</span>

<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">log_marginal</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute log-likelihood: P(Data | beta, M) - CORRECTED VECTORIZATION</span>
<span class="n">compute_log_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Compute all predictions: each row i gets genotypes[i] * beta</span>
<span class="w">  </span><span class="c1"># genotypes is Nx1, beta is length 2, predicted should be Nx2</span>
<span class="w">  </span><span class="n">predicted</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genotypes</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="w">  </span><span class="c1"># This gives N x 2 matrix correctly</span>
<span class="w">  </span><span class="n">residuals</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">phenotypes</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">predicted</span><span class="w">  </span><span class="c1"># N x 2 matrix</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Pre-compute for efficiency</span>
<span class="w">  </span><span class="n">residual_cov_inv</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">solve</span><span class="p">(</span><span class="n">residual_cov</span><span class="p">)</span>
<span class="w">  </span><span class="n">log_det_term</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">-0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">det</span><span class="p">(</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="kc">pi</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Compute quadratic form for each observation</span>
<span class="w">  </span><span class="n">log_lik</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">residuals</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="m">-0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">residual_cov_inv</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">r</span>
<span class="w">  </span><span class="p">}))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">N</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">log_det_term</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">log_lik</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute marginal likelihoods for all models and add to results dataframe</span>
<span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">),</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes_scaled</span><span class="p">,</span><span class="w"> </span><span class="n">Sigma_residual</span><span class="p">)</span>
<span class="p">})</span>

<span class="c1"># Add log marginal likelihoods to the dataframe</span>
<span class="n">results_df</span><span class="o">$</span><span class="n">Log_ML</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nLog Marginal Likelihoods:\n&quot;</span><span class="p">)</span>
<span class="n">results_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Log Marginal Likelihoods:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 7 × 3</caption>
<thead>
	<tr><th scope=col>Model</th><th scope=col>Prior</th><th scope=col>Log_ML</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>M0</td><td>0.1428571</td><td> -728.47</td></tr>
	<tr><td>M1</td><td>0.1428571</td><td>-3966.54</td></tr>
	<tr><td>M2</td><td>0.1428571</td><td>-3754.45</td></tr>
	<tr><td>M3</td><td>0.1428571</td><td> -498.04</td></tr>
	<tr><td>M4</td><td>0.1428571</td><td> -461.11</td></tr>
	<tr><td>M5</td><td>0.1428571</td><td> -459.43</td></tr>
	<tr><td>M6</td><td>0.1428571</td><td> -462.80</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
<section id="id3">
<h3>Posterior Distribution<a class="headerlink" href="#id3" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># ============================================================================</span>
<span class="c1"># Posterior Probabilities and Model Comparison</span>
<span class="c1"># ============================================================================</span>

<span class="c1"># Step 1: Compute unnormalized posterior (log scale)</span>
<span class="c1"># Posterior ∝ Likelihood × Prior</span>
<span class="n">log_posterior_unnorm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">results_df</span><span class="o">$</span><span class="n">Prior</span><span class="p">)</span>

<span class="c1"># Step 2: Normalize to obtain proper posterior probabilities</span>
<span class="c1"># Use log-sum-exp trick for numerical stability</span>
<span class="n">max_log_post</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_posterior_unnorm</span><span class="p">)</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_posterior_unnorm</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log_post</span><span class="p">)</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">posterior_probs</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)</span>

<span class="c1"># Step 3: Calculate Bayes factors relative to null model (M0)</span>
<span class="c1"># BF &gt; 1 indicates evidence in favor of the model vs null</span>
<span class="n">log_bayes_factors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="p">[</span><span class="s">&quot;M0&quot;</span><span class="p">]</span>
<span class="n">bayes_factors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_bayes_factors</span><span class="p">)</span>

<span class="c1"># Add posterior probabilities and Bayes factors to the dataframe</span>
<span class="n">results_df</span><span class="o">$</span><span class="n">Posterior</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)</span>
<span class="n">results_df</span><span class="o">$</span><span class="n">BF_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">bayes_factors</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="c1"># ============================================================================</span>
<span class="c1"># Summary Results</span>
<span class="c1"># ============================================================================</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nBayesian Model Comparison Results:\n&quot;</span><span class="p">)</span>
<span class="n">results_df</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\n\nMost probable model:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">results_df</span><span class="o">$</span><span class="n">Model</span><span class="p">[</span><span class="nf">which.max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)])</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nPosterior probability:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Bayesian Model Comparison Results:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 7 × 5</caption>
<thead>
	<tr><th scope=col>Model</th><th scope=col>Prior</th><th scope=col>Log_ML</th><th scope=col>Posterior</th><th scope=col>BF_vs_M0</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>M0</td><td>0.1428571</td><td> -728.47</td><td>0.0000</td><td> 1.000000e+00</td></tr>
	<tr><td>M1</td><td>0.1428571</td><td>-3966.54</td><td>0.0000</td><td> 0.000000e+00</td></tr>
	<tr><td>M2</td><td>0.1428571</td><td>-3754.45</td><td>0.0000</td><td> 0.000000e+00</td></tr>
	<tr><td>M3</td><td>0.1428571</td><td> -498.04</td><td>0.0000</td><td>1.185428e+100</td></tr>
	<tr><td>M4</td><td>0.1428571</td><td> -461.11</td><td>0.1527</td><td>1.300474e+116</td></tr>
	<tr><td>M5</td><td>0.1428571</td><td> -459.43</td><td>0.8194</td><td>6.978252e+116</td></tr>
	<tr><td>M6</td><td>0.1428571</td><td> -462.80</td><td>0.0279</td><td>2.380039e+115</td></tr>
</tbody>
</table>
</div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Most probable model: M5
Posterior probability: 0.8194 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">options</span><span class="p">(</span><span class="n">repr.plot.width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="n">repr.plot.height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">)</span>
<span class="c1"># Create plotting data with both prior and posterior from results_df</span>
<span class="n">plot_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">results_df</span><span class="o">$</span><span class="n">Model</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w">  </span><span class="n">Probability</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">results_df</span><span class="o">$</span><span class="n">Prior</span><span class="p">,</span><span class="w"> </span><span class="n">results_df</span><span class="o">$</span><span class="n">Posterior</span><span class="p">),</span>
<span class="w">  </span><span class="n">Type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">factor</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Prior&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Posterior&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">results_df</span><span class="p">)),</span><span class="w"> </span>
<span class="w">                </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Prior&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Posterior&quot;</span><span class="p">))</span>
<span class="p">)</span>

<span class="c1"># Create barplot with prior and posterior side-by-side</span>
<span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">plot_data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Model</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Probability</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Type</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_col</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_dodge</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">Probability</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span>
<span class="w">            </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_dodge</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">fontface</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Prior&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;tomato&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Posterior&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;darkgreen&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span>
<span class="w">    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Bayesian Model Comparison: Prior vs Posterior&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Probability&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Model&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylim</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span>
<span class="w">    </span><span class="n">plot.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span>
<span class="w">  </span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/e0ff26829b039ec9528b07598a275069f5b65c048719328cbf4525b4387ecfcc.png"><img alt="_images/e0ff26829b039ec9528b07598a275069f5b65c048719328cbf4525b4387ecfcc.png" src="_images/e0ff26829b039ec9528b07598a275069f5b65c048719328cbf4525b4387ecfcc.png" style="width: 960px; height: 720px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">ggsave</span><span class="p">(</span><span class="s">&quot;./figures/Bayesian_model_comparison.png&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Bayesian_multivariate_normal_mean_model.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bayesian Multivariate Normal Mean Model</p>
      </div>
    </a>
    <a class="right-next"
       href="Bayesian_mixture_model.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bayesian Mixture Model</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Bayesian Model Comparison</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bayes-factor">Bayes Factor</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-odds">Posterior Odds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-the-marginal-likelihood">Computing the Marginal Likelihood</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-fine-mapping">Example 1 – Fine Mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#prior-distribution">Prior Distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihoods-and-bayes-factors">Computing Marginal Likelihoods and Bayes Factors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-distribution">Posterior Distribution</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-detecting-pleiotropy">Example 2 – Detecting pleiotropy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Prior Distribution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihood-and-bayes-factor">Computing Marginal Likelihood and Bayes Factor</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Posterior Distribution</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gao Wang and Rui Dong
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>