
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Bayesian Model Comparison &#8212; Statistical backgrounds for genetics trainees</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Bayesian_model_comparison';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bayesian Mixture Model" href="Bayesian_mixture_model.html" />
    <link rel="prev" title="Bayesian Multivariate Normal Mean Model" href="Bayesian_multivariate_normal_mean_model.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="README.html">
  
  
  
  
  
  
    <p class="title logo__title">Statistical backgrounds for genetics trainees</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Popular Topics and Related Notebooks</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="notations.html">Notations</a></li>
<li class="toctree-l1"><a class="reference internal" href="popular_topics.html">Popular Topics in Statistical Genetics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Genetics Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genotype_coding.html">Genotype Coding</a></li>







<li class="toctree-l1"><a class="reference internal" href="minor_allele_frequency.html">Minor Allele Frequency</a></li>





<li class="toctree-l1"><a class="reference internal" href="Hardy_Weinberg_equilibrium.html">Hardy-Weinberg Equilibrium</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Variants</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium.html">Linkage Disequilibrium</a></li>






<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium_score.html">Linkage Disequilibrium Score</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Individuals</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_relationship_matrix.html">Genetic Relationship Matrix</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Fixed Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ordinary_least_squares.html">Ordinary Least Squares</a></li>






<li class="toctree-l1"><a class="reference internal" href="odds_ratio.html">Odds Ratio</a></li>






<li class="toctree-l1"><a class="reference internal" href="marginal_joint_effects.html">Marginal and Joint Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="summary_statistics.html">Summary Statistics</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Random Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="proportion_of_variance_explained.html">Proportion of Variance Explained and Heritability</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Linear Mixed Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linear_mixed_model.html">Linear Mixed Model</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Covariates</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="confounder.html">Confounder</a></li>





<li class="toctree-l1"><a class="reference internal" href="collider.html">Collider</a></li>





<li class="toctree-l1"><a class="reference internal" href="mediator.html">Mediator</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Meta-analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="meta_analysis_fixed_effect.html">Meta-Analysis Fixed Effect</a></li>





<li class="toctree-l1"><a class="reference internal" href="meta_analysis_random_effect.html">Meta-Analysis Random Effect</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Likelihood</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="likelihood.html">Likelihood</a></li>






<li class="toctree-l1"><a class="reference internal" href="maximum_likelihood_estimation.html">Maximum Likelihood Estimation</a></li>






<li class="toctree-l1"><a class="reference internal" href="likelihood_ratio.html">Likelihood Ratio</a></li>







<li class="toctree-l1"><a class="reference internal" href="expectation_maximum.html">Expectation-Maximum Algorithm</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Versus Frequentist</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_versus_frequentist.html">Bayesian versus Frequentist</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bayes_rule.html">Bayes Rule</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayes_factor.html">Bayes Factor</a></li>





<li class="toctree-l1"><a class="reference internal" href="p_value_and_Bayesian_hypothesis_testing.html">p-value and Bayesian Hypothesis Testing</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_normal_mean_model.html">Bayesian Normal Mean Model</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_multivariate_normal_mean_model.html">Bayesian Multivariate Normal Mean Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiple Bayesian Models</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Bayesian Model Comparison</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_mixture_model.html">Bayesian Mixture Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Latent Structures in Data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="factor_analysis.html">Factor Analysis</a></li>





<li class="toctree-l1"><a class="reference internal" href="principal_component_analysis.html">Principal Component Analysis</a></li>







<li class="toctree-l1"><a class="reference internal" href="hidden_Markov_model.html">Hidden Markov Model</a></li>






</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/Bayesian_model_comparison.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Bayesian Model Comparison</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Bayesian Model Comparison</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recall-bayes-factor">Recall Bayes Factor</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#from-uncertainty-quantification-to-model-selection">From Uncertainty Quantification to Model Selection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantifying-prior-model-uncertainty">Quantifying Prior Model Uncertainty</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-the-marginal-likelihood">Computing the Marginal Likelihood</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-case-when-we-can-condition-analytically">Simple Case: When We Can Condition Analytically</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#complex-case-when-parameter-uncertainty-cannot-be-resolved-analytically">Complex Case: When Parameter Uncertainty Cannot Be Resolved Analytically</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-fine-mapping">Example 1 – Fine Mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-simulated-data">Generation of Simulated Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-our-bayesian-framework">Setting Up Our Bayesian Framework</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihoods-and-bayes-factors">Computing Marginal Likelihoods and Bayes Factors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-posterior-model-probabilities">Computing Posterior Model Probabilities</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-detecting-pleiotropy">Example 2 – Detecting pleiotropy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-data-from-a-true-model">Generate Data from a True Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bayesian-model-comparison-setup">Bayesian Model Comparison Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marginal-likelihood-calculation">Marginal Likelihood Calculation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Bayesian Model Comparison</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-probabilities-and-results">Posterior Probabilities and Results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#supplementary">Supplementary</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Graphical Summary</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="bayesian-model-comparison">
<h1>Bayesian Model Comparison<a class="headerlink" href="#bayesian-model-comparison" title="Link to this heading">#</a></h1>
<p>Bayesian model comparison arises when we cannot condition on which model generated our data, leaving this model choice uncertain, so we <strong>quantify our uncertainty</strong> about competing models by measuring how well they predict observed data while accounting for the uncertainty inherent in their parameter spaces reflected by how broadly their priors spread probability.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="graphical-summary">
<h1>Graphical Summary<a class="headerlink" href="#graphical-summary" title="Link to this heading">#</a></h1>
<p><img alt="Fig" src="_images/Slide28.png" /></p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="key-formula">
<h1>Key Formula<a class="headerlink" href="#key-formula" title="Link to this heading">#</a></h1>
<p>When we cannot condition on which model <span class="math notranslate nohighlight">\(M_1\)</span> or <span class="math notranslate nohighlight">\(M_2\)</span> generated our data, we quantify our uncertainty about model choice by calculating <strong>posterior odds</strong>. These represent our updated uncertainty about competing models after observing the data:</p>
<div class="math notranslate nohighlight">
\[ \frac{P(M_1|D)}{P(M_2|D)} = \frac{P(D|M_1)P(M_1)}{P(D|M_2)P(M_2)} \]</div>
<p>This can be decomposed as:</p>
<div class="math notranslate nohighlight">
\[ \text{Posterior Odds} = \text{Bayes Factor} \times \text{Prior Odds} \]</div>
<p>Where:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(P(M_1|D)\)</span> and <span class="math notranslate nohighlight">\(P(M_2|D)\)</span> are our <strong>posterior uncertainties</strong> about models <span class="math notranslate nohighlight">\(M_1\)</span> and <span class="math notranslate nohighlight">\(M_2\)</span> after conditioning on the observed data</p></li>
<li><p><span class="math notranslate nohighlight">\(\frac{P(D|M_1)}{P(D|M_2)}\)</span> is the <strong>Bayes factor</strong>, which quantifies how the data updates our uncertainty about which model is more plausible</p></li>
<li><p><span class="math notranslate nohighlight">\(P(M_1)\)</span> and <span class="math notranslate nohighlight">\(P(M_2)\)</span> are our <strong>prior uncertainties</strong> about each model before conditioning on any data, representing what we cannot know a priori about model choice</p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="technical-details">
<h1>Technical Details<a class="headerlink" href="#technical-details" title="Link to this heading">#</a></h1>
<section id="recall-bayes-factor">
<h2>Recall Bayes Factor<a class="headerlink" href="#recall-bayes-factor" title="Link to this heading">#</a></h2>
<p><strong>Bayes factor</strong> quantifies how the data updates our uncertainty about competing models when we cannot condition on which model generated the observations:</p>
<div class="math notranslate nohighlight">
\[
\text{BF}_{21} = \frac{L(\text{M}_2|\text{D})}{L(\text{M}_1|\text{D})} ={\frac {\int \Pr(\theta_{2}|M_{2})\Pr(D|\theta_{2},M_{2})\,d\theta _{2}}{\int \Pr(\theta_{1}|M_{1})\Pr(D|\theta_{1},M_{1})\,d\theta _{1}}}
\]</div>
<p><strong>Interpreting uncertainty updates</strong>:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} &gt; 1\)</span>: Data reduces our uncertainty by favoring model <span class="math notranslate nohighlight">\(M_2\)</span> over <span class="math notranslate nohighlight">\(M_1\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} &lt; 1\)</span>: Data reduces our uncertainty by favoring model <span class="math notranslate nohighlight">\(M_1\)</span> over <span class="math notranslate nohighlight">\(M_2\)</span></p></li>
<li><p><span class="math notranslate nohighlight">\(\text{BF}_{21} = 1\)</span>: Data provides no information to reduce our uncertainty between models</p></li>
</ul>
</section>
<section id="from-uncertainty-quantification-to-model-selection">
<h2>From Uncertainty Quantification to Model Selection<a class="headerlink" href="#from-uncertainty-quantification-to-model-selection" title="Link to this heading">#</a></h2>
<p>Since we cannot condition on the true model a priori, the Bayes factor updates our prior uncertainty through:</p>
<div class="math notranslate nohighlight">
\[
\frac{P(M_2|D)}{P(M_1|D)} = \text{BF}_{21} \times \frac{P(M_2)}{P(M_1)}
\]</div>
<p>This shows how conditioning on data transforms our uncertainty:</p>
<div class="math notranslate nohighlight">
\[
\text{Posterior Uncertainty} = \text{Evidence Update} \times \text{Prior Uncertainty}
\]</div>
</section>
<section id="quantifying-prior-model-uncertainty">
<h2>Quantifying Prior Model Uncertainty<a class="headerlink" href="#quantifying-prior-model-uncertainty" title="Link to this heading">#</a></h2>
<p><strong>Equal uncertainty</strong> (no initial preference): Set <span class="math notranslate nohighlight">\(P(M_1) = P(M_2) = 0.5\)</span></p>
<div class="math notranslate nohighlight">
\[
\frac{P(M_2|D)}{P(M_1|D)} = \text{BF}_{21}
\]</div>
<p><em>Our posterior uncertainty is determined entirely by the data.</em></p>
<p><strong>Unequal uncertainty</strong> (one model initially preferred):</p>
<div class="math notranslate nohighlight">
\[
\frac{P(M_2|D)}{P(M_1|D)} = \text{BF}_{21} \times \frac{P(M_2)}{P(M_1)}
\]</div>
<p><em>The data updates our initial uncertainty about model choice.</em></p>
</section>
<section id="computing-the-marginal-likelihood">
<h2>Computing the Marginal Likelihood<a class="headerlink" href="#computing-the-marginal-likelihood" title="Link to this heading">#</a></h2>
<p>The integral <span class="math notranslate nohighlight">\(\int \Pr(\theta|M)\Pr(D|\theta,M)\,d\theta\)</span> represents the <strong>marginal likelihood</strong> - the probability of observing the data under a model when we cannot condition on specific parameter values and must average over all possibilities.</p>
<section id="simple-case-when-we-can-condition-analytically">
<h3>Simple Case: When We Can Condition Analytically<a class="headerlink" href="#simple-case-when-we-can-condition-analytically" title="Link to this heading">#</a></h3>
<p>When the prior and likelihood are conjugate (like the normal distributions in our Bayesian normal mean and multivariate normal mean models), we can analytically integrate out the parameter uncertainty, yielding closed-form marginal likelihoods.</p>
</section>
<section id="complex-case-when-parameter-uncertainty-cannot-be-resolved-analytically">
<h3>Complex Case: When Parameter Uncertainty Cannot Be Resolved Analytically<a class="headerlink" href="#complex-case-when-parameter-uncertainty-cannot-be-resolved-analytically" title="Link to this heading">#</a></h3>
<p>When priors and likelihoods are not conjugate, we may not be able to analytically resolve the parameter uncertainty in <span class="math notranslate nohighlight">\(\int \Pr(\theta|M)\Pr(D|\theta,M)\,d\theta\)</span>. This reflects the fundamental challenge: when we cannot condition on parameters, we must numerically integrate over all possible values, which becomes computationally prohibitive for high-dimensional parameter spaces in real GWAS datasets.</p>
<p>Several approximation methods have been developed to handle this uncertainty:</p>
<ul class="simple">
<li><p><strong>Laplace approximation</strong> (assumes uncertainty is concentrated around the posterior mode)</p></li>
<li><p><strong>MCMC-based methods</strong> (e.g., Chib’s method or thermodynamic integration, which samples from the uncertainty distribution)</p></li>
<li><p><strong>Nested Sampling</strong> (systematically explores the uncertainty landscape)</p></li>
</ul>
<p>The choice reflects a trade-off between computational feasibility and how thoroughly we can characterize the parameter uncertainty in our model comparison.</p>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="related-topics">
<h1>Related Topics<a class="headerlink" href="#related-topics" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/ordinary_least_squares.html">ordinary least squares</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/odds_ratio.html">odds ratio</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/summary_statistics.html">summary statistics</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/linkage_disequilibrium.html">linkage disequilibrium</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/marginal_joint_effects.html">marginal and joint effects</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_factor.html">Bayes factor</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_rule.html">Bayes rule</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_normal_mean_model.html">Bayesian normal mean model</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_multivariate_normal_mean_model.html">Bayesian multivariate normal mean model</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_mixture_model.html">Bayesian mixture model</a></p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="example">
<h1>Example<a class="headerlink" href="#example" title="Link to this heading">#</a></h1>
<section id="example-1-fine-mapping">
<h2>Example 1 – Fine Mapping<a class="headerlink" href="#example-1-fine-mapping" title="Link to this heading">#</a></h2>
<p>Remember when we discussed marginal versus joint effects? We saw how correlated predictors create misleading associations, and our solution was simple: fit a joint model with all predictors. This approach treated effects as fixed parameters, assuming we could condition on everything simultaneously.</p>
<p>But this assumption is not realistic - we cannot condition on everything. In genomics, when you have hundreds of correlated variants in linkage disequilibrium (LD), we cannot condition on all joint effects because the models become unstable due to overfitting, forcing us to acknowledge our conditioning limitations.</p>
<p>This creates the classic <strong>fine-mapping</strong> problem - because we cannot condition on everything, we face fundamental uncertainty about which variant is truly causal. Variants in LD “travel together” through generations, leaving us uncertain about causality even though all variants show association.</p>
<p>Since we cannot condition on everything, we must treat causal variant identity as uncertain. Bayesian model comparison provides a principled framework for quantifying our uncertainty about different causal hypotheses, systematically evaluating which single variant is most likely causal while properly accounting for this inherent uncertainty arising from our conditioning limitations.</p>
<p>Let’s define our competing models:</p>
<ul class="simple">
<li><p><strong>M0</strong>: Null model - none of the variants are causal</p></li>
<li><p><strong>M1</strong>: Variant 1 is the causal variant</p></li>
<li><p><strong>M2</strong>: Variant 2 is the causal variant</p></li>
<li><p><strong>M3</strong>: Variant 3 is the causal variant</p></li>
</ul>
<section id="generation-of-simulated-data">
<h3>Generation of Simulated Data<a class="headerlink" href="#generation-of-simulated-data" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Clear the environment and set seed for reproducibility</span>
<span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">9</span><span class="p">)</span><span class="w"> </span>

<span class="c1"># Parameters for our simulation</span>
<span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">20</span><span class="w">  </span><span class="c1"># Number of individuals</span>
<span class="n">M</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">3</span><span class="w">   </span><span class="c1"># Number of variants</span>

<span class="c1"># Generate correlated genotype data to simulate linkage disequilibrium</span>
<span class="c1"># We&#39;ll make Variant 1 the true causal variant</span>

<span class="c1"># Start with variant 1 (our true causal variant)</span>
<span class="n">variant1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="m">0.4</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">))</span>

<span class="c1"># Create variants 2 and 3 in LD with variant 1</span>
<span class="c1"># Higher correlation = stronger LD</span>
<span class="n">variant2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">runif</span><span class="p">(</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="n">variant3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">runif</span><span class="p">(</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.8</span><span class="p">,</span><span class="w"> </span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>

<span class="c1"># Combine into genotype matrix</span>
<span class="n">Xraw_additive</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">variant1</span><span class="p">,</span><span class="w"> </span><span class="n">variant2</span><span class="p">,</span><span class="w"> </span><span class="n">variant3</span><span class="p">)</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Individual&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">N</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Variant&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">M</span><span class="p">)</span>

<span class="c1"># Standardize genotypes (mean 0, variance 1)</span>
<span class="n">X</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">Xraw_additive</span><span class="p">,</span><span class="w"> </span><span class="n">center</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>

<span class="c1"># Generate phenotype where ONLY Variant 1 has a true causal effect</span>
<span class="n">true_beta1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1.5</span><span class="w">  </span><span class="c1"># Strong effect size for variant 1</span>
<span class="n">epsilon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rnorm</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">  </span><span class="c1"># Random noise</span>
<span class="n">Y_raw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">X</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">epsilon</span><span class="w">   </span><span class="c1"># Only variant 1 affects the trait</span>

<span class="c1"># Standardize phenotype</span>
<span class="n">Y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">Y_raw</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Notice how we’ve created linkage disequilibrium - variants 2 and 3 are correlated with variant 1, but only variant 1 actually causes the phenotype.</p>
</section>
<section id="setting-up-our-bayesian-framework">
<h3>Setting Up Our Bayesian Framework<a class="headerlink" href="#setting-up-our-bayesian-framework" title="Link to this heading">#</a></h3>
<p>Now we need to define our statistical models. For each model, we’ll assume:</p>
<ul class="simple">
<li><p><strong>Likelihood</strong>: Normal distribution for the phenotype</p></li>
<li><p><strong>Prior for effect size</strong>: Normal distribution centered at 0</p></li>
<li><p><strong>Prior for error variance</strong>: Inverse-gamma distribution</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Bayesian model setup</span>
<span class="c1"># We&#39;ll use conjugate priors for computational convenience</span>

<span class="c1"># Prior parameters for effect sizes (beta)</span>
<span class="n">beta_prior_mean</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="n">beta_prior_var</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span>

<span class="c1"># Prior parameters for error variance (sigma^2)</span>
<span class="c1"># Using inverse-gamma: IG(alpha, beta)</span>
<span class="n">sigma2_prior_alpha</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span>
<span class="n">sigma2_prior_beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="computing-marginal-likelihoods-and-bayes-factors">
<h3>Computing Marginal Likelihoods and Bayes Factors<a class="headerlink" href="#computing-marginal-likelihoods-and-bayes-factors" title="Link to this heading">#</a></h3>
<p>The key to Bayesian model comparison is computing the marginal likelihood for each model. This tells us how well each model explains the data, accounting for uncertainty in the parameters.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to compute marginal likelihood for each single-variant model</span>
<span class="n">compute_marginal_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">X_variant</span><span class="p">,</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Posterior parameters (using conjugate normal-inverse-gamma)</span>
<span class="w">  </span><span class="n">V_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">X_variant</span><span class="o">^</span><span class="m">2</span><span class="p">))</span>
<span class="w">  </span><span class="n">mu_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">V_n</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="n">beta_prior_mean</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">X_variant</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">Y</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">alpha_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sigma2_prior_alpha</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">n</span><span class="o">/</span><span class="m">2</span>
<span class="w">  </span><span class="n">beta_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sigma2_prior_beta</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">Y</span><span class="o">^</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">                                       </span><span class="n">beta_prior_mean</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">                                       </span><span class="n">mu_n</span><span class="o">^</span><span class="m">2</span><span class="o">/</span><span class="n">V_n</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Marginal likelihood calculation</span>
<span class="w">  </span><span class="n">log_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="o">-</span><span class="n">n</span><span class="o">/</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="kc">pi</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">            </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">V_n</span><span class="o">/</span><span class="n">beta_prior_var</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">            </span><span class="n">sigma2_prior_alpha</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">sigma2_prior_beta</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">            </span><span class="n">alpha_n</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">beta_n</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">            </span><span class="nf">lgamma</span><span class="p">(</span><span class="n">alpha_n</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nf">lgamma</span><span class="p">(</span><span class="n">sigma2_prior_alpha</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">log_ml</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Compute marginal likelihoods for each model</span>
<span class="n">log_ml_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">N</span><span class="p">),</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">  </span><span class="c1"># Null model</span>
<span class="n">log_ml_M1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">     </span><span class="c1"># Variant 1 causal</span>
<span class="n">log_ml_M2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">     </span><span class="c1"># Variant 2 causal  </span>
<span class="n">log_ml_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span><span class="n">X</span><span class="p">[,</span><span class="m">3</span><span class="p">],</span><span class="w"> </span><span class="n">Y</span><span class="p">)</span><span class="w">     </span><span class="c1"># Variant 3 causal</span>

<span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">log_ml_M0</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M1</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="p">,</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;M0 (Null)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M1 (Var1)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 (Var2)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M3 (Var3)&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can compute Bayes factors to compare models. Remember: BF &gt; 1 means the numerator model is preferred.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute Bayes factors relative to the null model</span>
<span class="n">BF_M1_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>
<span class="n">BF_M2_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>
<span class="n">BF_M3_vs_M0</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M3</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M0</span><span class="p">)</span>

<span class="c1"># Compute Bayes factors between variant models</span>
<span class="n">BF_M1_vs_M2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M2</span><span class="p">)</span>
<span class="n">BF_M1_vs_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>
<span class="n">BF_M2_vs_M3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_ml_M2</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_ml_M3</span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Bayes Factors (evidence in favor of first model):&quot;</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M1 vs M0 (Null):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M1_vs_M0</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M2 vs M0 (Null):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M2_vs_M0</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M3 vs M0 (Null):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M3_vs_M0</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
<span class="nf">print</span><span class="p">(</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Comparing variant models:&quot;</span><span class="p">)</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M1 vs M2:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M1_vs_M2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M1 vs M3:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M1_vs_M3</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;M2 vs M3:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">BF_M2_vs_M3</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Bayes Factors (evidence in favor of first model):&quot;
[1] &quot;M1 vs M0 (Null): 6783402.04&quot;
[1] &quot;M2 vs M0 (Null): 4.06&quot;
[1] &quot;M3 vs M0 (Null): 7763.49&quot;
[1] &quot;&quot;
[1] &quot;Comparing variant models:&quot;
[1] &quot;M1 vs M2: 1671050.19&quot;
[1] &quot;M1 vs M3: 873.76&quot;
[1] &quot;M2 vs M3: 0&quot;
</pre></div>
</div>
</div>
</div>
</section>
<section id="computing-posterior-model-probabilities">
<h3>Computing Posterior Model Probabilities<a class="headerlink" href="#computing-posterior-model-probabilities" title="Link to this heading">#</a></h3>
<p>Finally, let’s compute the posterior probability for each model. This tells us how confident we should be in each explanation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Assume equal prior probabilities for all models</span>
<span class="n">prior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0.25</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">  </span><span class="c1"># Equal priors for M0, M1, M2, M3</span>

<span class="c1"># Compute posterior probabilities</span>
<span class="c1"># First, convert log marginal likelihoods to regular scale (avoiding overflow)</span>
<span class="n">max_log_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">)</span>
<span class="n">scaled_ml</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log_ml</span><span class="p">)</span>

<span class="c1"># Compute posterior probabilities</span>
<span class="n">unnormalized_posterior</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prior_probs</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">scaled_ml</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unnormalized_posterior</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">unnormalized_posterior</span><span class="p">)</span>

<span class="nf">names</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;M0 (Null)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M1 (Var1)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M2 (Var2)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;M3 (Var3)&quot;</span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="s">&quot;Posterior Model Probabilities:&quot;</span><span class="p">)</span>
<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="m">4</span><span class="p">)))</span>
<span class="p">}</span>

<span class="c1"># Find the most probable model</span>
<span class="n">best_model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)[</span><span class="nf">which.max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)]</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Most probable model:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">best_model</span><span class="p">))</span>
<span class="nf">print</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Posterior probability:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[1] &quot;Posterior Model Probabilities:&quot;
[1] &quot;M0 (Null) : 0&quot;
[1] &quot;M1 (Var1) : 0.9989&quot;
[1] &quot;M2 (Var2) : 0&quot;
[1] &quot;M3 (Var3) : 0.0011&quot;
[1] &quot;Most probable model: M1 (Var1)&quot;
[1] &quot;Posterior probability: 0.9989&quot;
</pre></div>
</div>
</div>
</div>
<p>In this example, we demonstrated how Bayesian model comparison can solve the fine-mapping problem. Even when variants are in linkage disequilibrium and all show association with the phenotype, our method was able to identify the true causal variant by:</p>
<ol class="arabic simple">
<li><p><strong>Comparing multiple models</strong> rather than just testing individual associations</p></li>
<li><p><strong>Accounting for uncertainty</strong> in our parameter estimates</p></li>
<li><p><strong>Using proper statistical principles</strong> to weigh evidence for different explanations</p></li>
</ol>
<p>The key insight is that the Bayes factor naturally accounts for model complexity - a model needs to provide substantially better fit to the data to overcome the “penalty” for additional parameters. This helps us distinguish between true causation and mere correlation due to LD.</p>
<p>This approach forms the foundation for more sophisticated fine-mapping methods used in modern genomics research.</p>
</section>
</section>
<section id="example-2-detecting-pleiotropy">
<h2>Example 2 – Detecting pleiotropy<a class="headerlink" href="#example-2-detecting-pleiotropy" title="Link to this heading">#</a></h2>
<p>Recall our <a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayesian_multivariate_normal_mean_model.html">Lecture: Bayesian multivariate normal mean model</a> example where we studied the scenario that one genetic variant could affect two traits (height and weight).</p>
<p>In practice, there could be multiple options for the behavior of this variant on the two traits:</p>
<ul class="simple">
<li><p><strong>M0 (Null)</strong>: “This variant does nothing” - <span class="math notranslate nohighlight">\(\boldsymbol{\beta}\)</span> = (0, 0)</p></li>
<li><p><strong>M1 (Height Only)</strong>: “This variant only affects height” - <span class="math notranslate nohighlight">\(\boldsymbol{\beta}\)</span> = (<span class="math notranslate nohighlight">\(\beta_1\)</span>, 0)</p></li>
<li><p><strong>M2 (Weight Only)</strong>: “This variant only affects weight” - <span class="math notranslate nohighlight">\(\boldsymbol{\beta}\)</span> = (0, <span class="math notranslate nohighlight">\(\beta_2\)</span>)</p></li>
<li><p><strong>M3 (Perfect Correlation)</strong>: “Effects are perfectly correlated” - <span class="math notranslate nohighlight">\(\beta_2\)</span> = <span class="math notranslate nohighlight">\(\beta_1\)</span></p></li>
<li><p><strong>M4 (Weak Correlation)</strong>: “Effects are weakly related” - correlation = 0.1</p></li>
<li><p><strong>M5 (Medium Correlation)</strong>: “Effects are moderately related” - correlation = 0.5</p></li>
<li><p><strong>M6 (Strong Correlation)</strong>: “Effects are strongly related” - correlation = 0.8</p></li>
</ul>
<p>The true effect may come from one of the models above (or maybe some other models), but let’s use the Bayesian model comparison to detect which one is the most likely, conditioning on the data that we observe.</p>
<p>This is where Bayesian model comparison becomes powerful - it allows us to quantify our uncertainty when we cannot condition on the true biological model.</p>
<p><strong>Our Approach</strong>: We’ll use Bayesian model comparison to quantify our uncertainty about these competing biological explanations, computing marginal likelihoods <span class="math notranslate nohighlight">\(P(\text{D}|\text{M})\)</span> for each mixture model by integrating over all possible parameter values when we cannot condition on their true values.</p>
<section id="generate-data-from-a-true-model">
<h3>Generate Data from a True Model<a class="headerlink" href="#generate-data-from-a-true-model" title="Link to this heading">#</a></h3>
<p>Let’s simulate data where we know the truth - the variant has a medium correlation effect (M5).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">library</span><span class="p">(</span><span class="n">MASS</span><span class="p">)</span><span class="w">  </span><span class="c1"># for mvrnorm</span>
<span class="nf">library</span><span class="p">(</span><span class="n">mvtnorm</span><span class="p">)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">28</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Data Generation for Bayesian Model Comparison</span>
<span class="c1"># One variant affecting two traits (height and weight) with correlated effects</span>

<span class="c1"># Parameters</span>
<span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span><span class="w">  </span><span class="c1"># sample size</span>
<span class="n">true_correlation</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.5</span><span class="w">  </span><span class="c1"># correlation between effects on height and weight</span>
<span class="n">effect_variance</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.25</span><span class="w">  </span><span class="c1"># variance for each effect</span>

<span class="c1"># Step 1: Generate effect sizes from multivariate normal distribution</span>
<span class="c1"># Covariance matrix for effects with correlation = 0.5</span>
<span class="n">Sigma_effects</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">true_correlation</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span>
<span class="w">                         </span><span class="n">true_correlation</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="c1"># Draw the true effects from multivariate normal</span>
<span class="n">beta_true</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mvrnorm</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">mu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="n">Sigma</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sigma_effects</span><span class="p">)</span>
<span class="n">true_beta_height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">beta_true</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w">  </span><span class="c1"># effect on height</span>
<span class="n">true_beta_weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">beta_true</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w">  </span><span class="c1"># effect on weight</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Generated effect sizes:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Effect on height (beta1):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">true_beta_height</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Effect on weight (beta2):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">true_beta_weight</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>

<span class="c1"># Step 2: Generate genotype data</span>
<span class="c1"># Genotype coding: 0, 1, 2 copies of effect allele</span>
<span class="n">minor_allele_freq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.3</span>
<span class="n">genotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbinom</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">minor_allele_freq</span><span class="p">)</span>

<span class="c1"># Step 3: Generate phenotype data</span>
<span class="c1"># Residual variance-covariance matrix (environmental effects)</span>
<span class="n">Sigma_residual</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1.0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">,</span><span class="w">   </span><span class="c1"># height variance = 1.0, covariance = 0.2</span>
<span class="w">                          </span><span class="m">0.2</span><span class="p">,</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">  </span><span class="c1"># weight variance = 0.8</span>

<span class="c1"># Generate residual errors</span>
<span class="n">residual_errors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mvrnorm</span><span class="p">(</span><span class="n">N</span><span class="p">,</span><span class="w"> </span><span class="n">mu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">Sigma</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sigma_residual</span><span class="p">)</span>

<span class="c1"># Generate phenotypes: Y = genetic effect + residual error</span>
<span class="c1"># Height and weight with baseline means</span>
<span class="n">baseline_height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">170</span><span class="w">  </span><span class="c1"># cm</span>
<span class="n">baseline_weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">70</span><span class="w">   </span><span class="c1"># kg</span>

<span class="n">height</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">baseline_height</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">genotypes</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta_height</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">residual_errors</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span>
<span class="n">weight</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">baseline_weight</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">genotypes</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">true_beta_weight</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">residual_errors</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span>

<span class="c1"># Combine phenotypes</span>
<span class="n">phenotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">height</span><span class="p">,</span><span class="w"> </span><span class="n">weight</span><span class="p">)</span>
<span class="n">phenotypes_scaled</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">scale</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">phenotypes_scaled</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Height&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Weight&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated effect sizes:
Effect on height (beta1): 1.192 
Effect on weight (beta2): 2.16 
</pre></div>
</div>
</div>
</div>
</section>
<section id="bayesian-model-comparison-setup">
<h3>Bayesian Model Comparison Setup<a class="headerlink" href="#bayesian-model-comparison-setup" title="Link to this heading">#</a></h3>
<p>We’ll define all possible models as prior covariance matrices for the effect sizes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define covariance matrices for each model</span>
<span class="n">models</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span>
<span class="w">  </span><span class="n">M0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">                    </span><span class="c1"># Null: beta = (0, 0)</span>
<span class="w">  </span><span class="n">M1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">      </span><span class="c1"># Height only</span>
<span class="w">  </span><span class="n">M2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">      </span><span class="c1"># Weight only</span>
<span class="w">  </span><span class="n">M3</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w">       </span><span class="c1"># Perfect correlation</span>
<span class="w">               </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w">  </span><span class="n">M4</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="c1"># Weak correlation (0.1)</span>
<span class="w">               </span><span class="m">0.1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w">  </span><span class="n">M5</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="c1"># Medium correlation (0.5)</span>
<span class="w">               </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w">  </span><span class="n">M6</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="m">0.8</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="c1"># Strong correlation (0.8)</span>
<span class="w">               </span><span class="m">0.8</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">,</span><span class="w"> </span><span class="n">effect_variance</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Equal prior probabilities for all models</span>
<span class="n">prior_weights</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="o">/</span><span class="m">7</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">prior_weights</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nPrior weights (equal for all models):&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">prior_weights</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Prior weights (equal for all models): 0.143 0.143 0.143 0.143 0.143 0.143 0.143 
</pre></div>
</div>
</div>
</div>
</section>
<section id="marginal-likelihood-calculation">
<h3>Marginal Likelihood Calculation<a class="headerlink" href="#marginal-likelihood-calculation" title="Link to this heading">#</a></h3>
<p>When we cannot condition on the true parameter values, we must integrate over all possibilities to compute marginal likelihoods. This integration captures our uncertainty about parameters within each model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to compute log-likelihood when we condition on specific beta values</span>
<span class="n">log_likelihood_variant</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">N</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">phenotypes</span><span class="p">)</span>
<span class="w">  </span><span class="n">ll</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="w">  </span><span class="n">baseline</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w">  </span><span class="c1"># Centered phenotypes (subtract population means)</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">N</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">pred</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">baseline</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">genotypes_variant</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">beta</span>
<span class="w">    </span><span class="n">diff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">phenotypes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">pred</span>
<span class="w">    </span><span class="n">ll</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ll</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="n">diff</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="nf">solve</span><span class="p">(</span><span class="n">residual_cov</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">diff</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">          </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">det</span><span class="p">(</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="kc">pi</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">))</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">ll</span><span class="p">))</span>
<span class="p">}</span>

<span class="c1"># Function to compute prior probability when we condition on specific beta values</span>
<span class="n">log_prior</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">component_cov_matrix</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">all</span><span class="p">(</span><span class="n">component_cov_matrix</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># For null model: prior probability is 1 if beta = (0,0), 0 otherwise</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">all</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.01</span><span class="p">),</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="kc">Inf</span><span class="p">))</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Add small regularization for numerical stability</span>
<span class="w">  </span><span class="n">cov_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">component_cov_matrix</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">diag</span><span class="p">(</span><span class="m">1e-6</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="w">  </span><span class="n">lp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">-0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">t</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="nf">solve</span><span class="p">(</span><span class="n">cov_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">beta</span><span class="w"> </span><span class="o">-</span><span class="w"> </span>
<span class="w">        </span><span class="m">0.5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">det</span><span class="p">(</span><span class="m">2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="kc">pi</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">cov_matrix</span><span class="p">))</span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">lp</span><span class="p">))</span>
<span class="p">}</span>

<span class="c1"># Function to compute marginal likelihood for each model</span>
<span class="n">compute_marginal_likelihood</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">component_cov</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">models</span><span class="p">[[</span><span class="n">model_name</span><span class="p">]]</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">model_name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;M0&quot;</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># Null model: we can condition on beta = (0, 0)</span>
<span class="w">    </span><span class="n">beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
<span class="w">    </span><span class="n">ll</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_likelihood_variant</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span>
<span class="w">    </span><span class="n">lp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_prior</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">component_cov</span><span class="p">)</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">ll</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">lp</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># For non-null models: cannot condition on beta, so we integrate</span>
<span class="w">  </span><span class="c1"># This integration reflects our uncertainty about parameter values</span>
<span class="w">  </span><span class="n">beta_range</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">-1.5</span><span class="p">,</span><span class="w"> </span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="n">length.out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span><span class="w">  </span><span class="c1"># Integration grid</span>
<span class="w">  </span><span class="n">grid</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">expand.grid</span><span class="p">(</span><span class="n">beta_height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_range</span><span class="p">,</span><span class="w"> </span><span class="n">beta_weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">beta_range</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">log_integrand_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">grid</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">grid</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">beta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">grid</span><span class="o">$</span><span class="n">beta_height</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">grid</span><span class="o">$</span><span class="n">beta_weight</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="w">    </span><span class="n">ll</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_likelihood_variant</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes_variant</span><span class="p">,</span><span class="w"> </span><span class="n">residual_cov</span><span class="p">)</span>
<span class="w">    </span><span class="n">lp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log_prior</span><span class="p">(</span><span class="n">beta</span><span class="p">,</span><span class="w"> </span><span class="n">component_cov</span><span class="p">)</span>
<span class="w">    </span><span class="n">log_integrand_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ll</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">lp</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Numerical integration using log-sum-exp for stability</span>
<span class="w">  </span><span class="n">max_log</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_integrand_values</span><span class="p">[</span><span class="nf">is.finite</span><span class="p">(</span><span class="n">log_integrand_values</span><span class="p">)])</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.finite</span><span class="p">(</span><span class="n">max_log</span><span class="p">))</span><span class="w"> </span><span class="kr">return</span><span class="p">(</span><span class="o">-</span><span class="kc">Inf</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="n">log_marginal</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">max_log</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="nf">exp</span><span class="p">(</span><span class="n">log_integrand_values</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log</span><span class="p">),</span><span class="w"> </span><span class="n">na.rm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="w">  </span><span class="n">grid_spacing</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="n">beta_range</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">beta_range</span><span class="p">[</span><span class="m">1</span><span class="p">])</span><span class="o">^</span><span class="m">2</span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">log_marginal</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">grid_spacing</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="id1">
<h3>Bayesian Model Comparison<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate marginal likelihood for each model</span>
<span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="m">7</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">)</span>

<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">7</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">model_name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span>
<span class="w">  </span><span class="n">log_marginal_likelihoods</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">compute_marginal_likelihood</span><span class="p">(</span>
<span class="w">    </span><span class="n">model_name</span><span class="p">,</span><span class="w"> </span><span class="n">genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">phenotypes_scaled</span><span class="p">,</span><span class="w"> </span><span class="n">Sigma_residual</span>
<span class="w">  </span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="posterior-probabilities-and-results">
<h3>Posterior Probabilities and Results<a class="headerlink" href="#posterior-probabilities-and-results" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate posterior probabilities using Bayes&#39; theorem</span>
<span class="c1"># Posterior is in proportion to Likelihood * Prior</span>
<span class="n">log_posterior_unnorm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">prior_weights</span><span class="p">)</span>

<span class="c1"># Normalize to get proper probabilities (using log-sum-exp for numerical stability)</span>
<span class="n">max_log_post</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">log_posterior_unnorm</span><span class="p">)</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_posterior_unnorm</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">max_log_post</span><span class="p">)</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">posterior_probs</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)</span>

<span class="c1"># Calculate Bayes factors (relative to M0)</span>
<span class="n">log_bayes_factors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">log_marginal_likelihoods</span><span class="p">[</span><span class="s">&quot;M0&quot;</span><span class="p">]</span>
<span class="n">bayes_factors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">log_bayes_factors</span><span class="p">)</span>

<span class="c1"># Create results summary</span>
<span class="n">results_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">),</span>
<span class="w">  </span><span class="n">Log_Marginal_Likelihood</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">log_marginal_likelihoods</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">),</span>
<span class="w">  </span><span class="n">Prior_Probability</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">prior_weights</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span>
<span class="w">  </span><span class="n">Posterior_Probability</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span>
<span class="w">  </span><span class="n">Bayes_Factor_vs_M0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">bayes_factors</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w">  </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The Bayesian results for M0 to M6 are:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">results_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 7 x 5</caption>
<thead>
	<tr><th></th><th scope=col>Model</th><th scope=col>Log_Marginal_Likelihood</th><th scope=col>Prior_Probability</th><th scope=col>Posterior_Probability</th><th scope=col>Bayes_Factor_vs_M0</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>M0</th><td>M0</td><td>-416.572</td><td>0.1429</td><td>0.0000</td><td>1.000000e+00</td></tr>
	<tr><th scope=row>M1</th><td>M1</td><td>-874.699</td><td>0.1429</td><td>0.0000</td><td>0.000000e+00</td></tr>
	<tr><th scope=row>M2</th><td>M2</td><td>-819.736</td><td>0.1429</td><td>0.0000</td><td>0.000000e+00</td></tr>
	<tr><th scope=row>M3</th><td>M3</td><td>-344.314</td><td>0.1429</td><td>0.0059</td><td>2.405518e+31</td></tr>
	<tr><th scope=row>M4</th><td>M4</td><td>-340.764</td><td>0.1429</td><td>0.2043</td><td>8.369311e+32</td></tr>
	<tr><th scope=row>M5</th><td>M5</td><td>-340.047</td><td>0.1429</td><td>0.4187</td><td>1.715549e+33</td></tr>
	<tr><th scope=row>M6</th><td>M6</td><td>-340.167</td><td>0.1429</td><td>0.3711</td><td>1.520367e+33</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Find the winning model</span>
<span class="n">winner_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which.max</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">)</span>
<span class="n">winner_name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">)[</span><span class="n">winner_idx</span><span class="p">]</span>
<span class="n">winner_prob</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">posterior_probs</span><span class="p">[</span><span class="n">winner_idx</span><span class="p">]</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nMost supported model:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">winner_name</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Posterior probability:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">winner_prob</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Bayes factor vs M0:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">bayes_factors</span><span class="p">[</span><span class="n">winner_idx</span><span class="p">],</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Most supported model: M5 
Posterior probability: 0.4187 
Bayes factor vs M0: 1.715549e+33 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot: Posterior probabilities</span>
<span class="nf">barplot</span><span class="p">(</span><span class="n">posterior_probs</span><span class="p">,</span><span class="w"> </span><span class="n">names.arg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">),</span><span class="w"> </span>
<span class="w">        </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Posterior Model Probabilities&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Probability&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Model&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">models</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">winner_name</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;lightblue&quot;</span><span class="p">),</span>
<span class="w">        </span><span class="n">las</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">cex.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/e4249189258b8a7847fbb73cda4e7afb4e05e043d10b940ac613956c60050113.png"><img alt="_images/e4249189258b8a7847fbb73cda4e7afb4e05e043d10b940ac613956c60050113.png" src="_images/e4249189258b8a7847fbb73cda4e7afb4e05e043d10b940ac613956c60050113.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="supplementary">
<h1>Supplementary<a class="headerlink" href="#supplementary" title="Link to this heading">#</a></h1>
<section id="id2">
<h2>Graphical Summary<a class="headerlink" href="#id2" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>

<span class="c1"># Basic model comparison data with unequal priors</span>
<span class="n">model_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w"> </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Model 1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Model 2&quot;</span><span class="p">),</span>
<span class="w"> </span><span class="n">Log_Marginal_Likelihood</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">-152.3</span><span class="p">,</span><span class="w"> </span><span class="m">-148.7</span><span class="p">),</span>
<span class="w"> </span><span class="n">Prior_Probability</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w">  </span><span class="c1"># Higher prior for Model 1</span>
<span class="p">)</span>

<span class="c1"># Compute Bayes factor and posterior probabilities</span>
<span class="n">bayes_factor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">model_data</span><span class="o">$</span><span class="n">Log_Marginal_Likelihood</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">model_data</span><span class="o">$</span><span class="n">Log_Marginal_Likelihood</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>
<span class="n">unnormalized_posteriors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">exp</span><span class="p">(</span><span class="n">model_data</span><span class="o">$</span><span class="n">Log_Marginal_Likelihood</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">model_data</span><span class="o">$</span><span class="n">Prior_Probability</span>
<span class="n">posterior_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unnormalized_posteriors</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">unnormalized_posteriors</span><span class="p">)</span>

<span class="c1"># Create plotting data</span>
<span class="n">plot_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w"> </span><span class="n">Model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Model 1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Model 2&quot;</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">),</span>
<span class="w"> </span><span class="n">Probability</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">model_data</span><span class="o">$</span><span class="n">Prior_Probability</span><span class="p">,</span><span class="w"> </span><span class="n">posterior_probs</span><span class="p">),</span>
<span class="w"> </span><span class="n">Type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Prior&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Posterior&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Create barplot</span>
<span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">plot_data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Model</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Probability</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Type</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">geom_col</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_dodge</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">Probability</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">)),</span><span class="w"> </span>
<span class="w">           </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_dodge</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">),</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">fontface</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Prior&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;tomato&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Posterior&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;skyblue&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">labs</span><span class="p">(</span>
<span class="w">   </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Bayesian Model Comparison&quot;</span><span class="p">,</span>
<span class="w">   </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Bayes Factor (Model 2 vs Model 1) =&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">bayes_factor</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)),</span>
<span class="w">   </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Probability&quot;</span><span class="p">,</span>
<span class="w">   </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Model&quot;</span><span class="p">,</span>
<span class="w">   </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Type&quot;</span>
<span class="w"> </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">ylim</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1.05</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">theme_minimal</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w"> </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">   </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">plot.subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">   </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bottom&quot;</span>
<span class="w"> </span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>

<span class="c1"># Save the plot</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="s">&quot;./cartoons/Bayesian_model_comparison.png&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p</span><span class="p">,</span>
<span class="w">      </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span>
<span class="w">      </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span>
<span class="w">      </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/7955501d1793fef1431fa6a21335b0bab6bd6c5e9bf96c00eb55a01e02f2542c.png"><img alt="_images/7955501d1793fef1431fa6a21335b0bab6bd6c5e9bf96c00eb55a01e02f2542c.png" src="_images/7955501d1793fef1431fa6a21335b0bab6bd6c5e9bf96c00eb55a01e02f2542c.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Bayesian_multivariate_normal_mean_model.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bayesian Multivariate Normal Mean Model</p>
      </div>
    </a>
    <a class="right-next"
       href="Bayesian_mixture_model.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bayesian Mixture Model</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Bayesian Model Comparison</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#recall-bayes-factor">Recall Bayes Factor</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#from-uncertainty-quantification-to-model-selection">From Uncertainty Quantification to Model Selection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quantifying-prior-model-uncertainty">Quantifying Prior Model Uncertainty</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-the-marginal-likelihood">Computing the Marginal Likelihood</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simple-case-when-we-can-condition-analytically">Simple Case: When We Can Condition Analytically</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#complex-case-when-parameter-uncertainty-cannot-be-resolved-analytically">Complex Case: When Parameter Uncertainty Cannot Be Resolved Analytically</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-fine-mapping">Example 1 – Fine Mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-simulated-data">Generation of Simulated Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-our-bayesian-framework">Setting Up Our Bayesian Framework</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-marginal-likelihoods-and-bayes-factors">Computing Marginal Likelihoods and Bayes Factors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#computing-posterior-model-probabilities">Computing Posterior Model Probabilities</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-detecting-pleiotropy">Example 2 – Detecting pleiotropy</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generate-data-from-a-true-model">Generate Data from a True Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#bayesian-model-comparison-setup">Bayesian Model Comparison Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#marginal-likelihood-calculation">Marginal Likelihood Calculation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Bayesian Model Comparison</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-probabilities-and-results">Posterior Probabilities and Results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#supplementary">Supplementary</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Graphical Summary</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gao Wang and Rui Dong
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>