
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Expectation-Maximum Algorithm &#8212; Statistical backgrounds for genetics trainees</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'expectation_maximum';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bayesian versus Frequentist" href="Bayesian_versus_frequentist.html" />
    <link rel="prev" title="Likelihood Ratio" href="likelihood_ratio.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="README.html">
  
  
  
  
  
  
    <p class="title logo__title">Statistical backgrounds for genetics trainees</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Popular Topics and Related Notebooks</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="notations.html">Notations</a></li>
<li class="toctree-l1"><a class="reference internal" href="popular_topics.html">Popular Topics in Statistical Genetics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Genetics Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genotype_coding.html">Genotype Coding</a></li>







<li class="toctree-l1"><a class="reference internal" href="minor_allele_frequency.html">Minor Allele Frequency</a></li>





<li class="toctree-l1"><a class="reference internal" href="Hardy_Weinberg_equilibrium.html">Hardy-Weinberg Equilibrium</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Variants</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium.html">Linkage Disequilibrium</a></li>






<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium_score.html">Linkage Disequilibrium Score</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Individuals</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_relationship_matrix.html">Genetic Relationship Matrix</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Fixed Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ordinary_least_squares.html">Ordinary Least Squares</a></li>






<li class="toctree-l1"><a class="reference internal" href="odds_ratio.html">Odds Ratio</a></li>






<li class="toctree-l1"><a class="reference internal" href="marginal_joint_effects.html">Marginal and Joint Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="summary_statistics.html">Summary Statistics</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Random Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="random_effects.html">Random Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="proportion_of_variance_explained.html">Proportion of Variance Explained and Heritability</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Linear Mixed Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linear_mixed_model.html">Linear Mixed Model</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Covariates</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="confounder.html">Confounder</a></li>





<li class="toctree-l1"><a class="reference internal" href="collider.html">Collider</a></li>






<li class="toctree-l1"><a class="reference internal" href="mediator.html">Mediator</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Meta-analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="meta_analysis_fixed_effect.html">Meta-Analysis Fixed Effect</a></li>





<li class="toctree-l1"><a class="reference internal" href="meta_analysis_random_effect.html">Meta-Analysis Random Effect</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Likelihood</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="likelihood.html">Likelihood</a></li>






<li class="toctree-l1"><a class="reference internal" href="maximum_likelihood_estimation.html">Maximum Likelihood Estimation</a></li>






<li class="toctree-l1"><a class="reference internal" href="likelihood_ratio.html">Likelihood Ratio</a></li>







<li class="toctree-l1 current active"><a class="current reference internal" href="#">Expectation-Maximum Algorithm</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Versus Frequentist</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_versus_frequentist.html">Bayesian versus Frequentist</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bayes_rule.html">Bayes Rule</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayes_factor.html">Bayes Factor</a></li>





<li class="toctree-l1"><a class="reference internal" href="p_value_and_Bayesian_hypothesis_testing.html">p-value and Bayesian Hypothesis Testing</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_normal_mean_model.html">Bayesian Normal Mean Model</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_multivariate_normal_mean_model.html">Bayesian Multivariate Normal Mean Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiple Bayesian Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_model_comparison.html">Bayesian Model Comparison</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_mixture_model.html">Bayesian Mixture Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Latent Structures in Data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="factor_analysis.html">Factor Analysis</a></li>





<li class="toctree-l1"><a class="reference internal" href="principal_component_analysis.html">Principal Component Analysis</a></li>







<li class="toctree-l1"><a class="reference internal" href="hidden_Markov_model.html">Hidden Markov Model</a></li>






</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/expectation_maximum.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Expectation-Maximum Algorithm</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Expectation-Maximum Algorithm</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#why-em-works-the-monotonic-increase-guarantee">Why EM Works: The Monotonic Increase Guarantee</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-key-decomposition">The Key Decomposition</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#practical-implementation">Practical Implementation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#e-step-computing-responsibilities-via-bayes-rule">E-Step: Computing Responsibilities via Bayes’ Rule</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#m-step-maximizing-the-q-function">M-Step: Maximizing the Q-Function</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-generate-simulated-data">Step 1: Generate Simulated Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-develop-the-em-algorithm">Step 2: Develop the EM Algorithm</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-setup">Model Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#error-model">Error Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hardy-weinberg-model">Hardy-Weinberg Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#em-algorithm-implementation">EM Algorithm Implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-perform-the-em-algorithm">Step 3: Perform the EM Algorithm</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-results-and-visualization">Step 4: Results and Visualization</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="expectation-maximum-algorithm">
<h1>Expectation-Maximum Algorithm<a class="headerlink" href="#expectation-maximum-algorithm" title="Link to this heading">#</a></h1>
<p>The EM algorithm breaks an impossible optimization problem into two manageable steps that feed into each other: first calculating what the hidden data probably looks like given current parameter estimates, then finding better parameters by treating those probabilistic guesses as if they were observations.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="graphical-summary">
<h1>Graphical Summary<a class="headerlink" href="#graphical-summary" title="Link to this heading">#</a></h1>
<p><img alt="Fig" src="_images/Slide33.png" /></p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="key-formula">
<h1>Key Formula<a class="headerlink" href="#key-formula" title="Link to this heading">#</a></h1>
<p>Given the statistical model which generates a set <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> of observed data, a set of unobserved latent data <span class="math notranslate nohighlight">\(\mathbf{Z}\)</span>, and a vector of unknown parameters <span class="math notranslate nohighlight">\(\boldsymbol{\theta}\)</span>, along with a likelihood function <span class="math notranslate nohighlight">\(L(\boldsymbol{\theta}; \mathbf{X}, \mathbf{Z}) = p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})\)</span>, the maximum likelihood estimate (MLE) of the unknown parameters is determined by maximizing the marginal likelihood of the observed data</p>
<div class="math notranslate nohighlight">
\[
L(\boldsymbol{\theta}; \mathbf{X}) = p(\mathbf{X} \mid \boldsymbol{\theta}) = \int p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta}) \, d\mathbf{Z} = \int p(\mathbf{X} \mid \mathbf{Z}, \boldsymbol{\theta}) p(\mathbf{Z} \mid \boldsymbol{\theta}) \, d\mathbf{Z}
\]</div>
<p>However, this quantity is often intractable since <span class="math notranslate nohighlight">\(\mathbf{Z}\)</span> is unobserved and the distribution of <span class="math notranslate nohighlight">\(\mathbf{Z}\)</span> is unknown before attaining <span class="math notranslate nohighlight">\(\boldsymbol{\theta}\)</span>.</p>
<p>The <strong>EM algorithm</strong> seeks to find the <strong>maximum likelihood estimate (MLE)</strong> of the marginal likelihood by iteratively applying these two steps:</p>
<p><strong>Expectation step (E step):</strong> Define <span class="math notranslate nohighlight">\(Q(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})\)</span> as the expected value of the log likelihood function of <span class="math notranslate nohighlight">\(\boldsymbol{\theta}\)</span>, i.e., <span class="math notranslate nohighlight">\(\ell(\boldsymbol{\theta})\)</span>, with respect to the current conditional distribution of <span class="math notranslate nohighlight">\(\mathbf{Z}\)</span> given <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> and the current estimates of the parameters <span class="math notranslate nohighlight">\(\boldsymbol{\theta}^{(t)}\)</span>:</p>
<div class="math notranslate nohighlight">
\[
Q(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)}) = \mathbb{E}_{\mathbf{Z} \sim p(\cdot \mid \mathbf{X}, \boldsymbol{\theta}^{(t)})} [\log p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta})] := \int \log p(\mathbf{X}, \mathbf{Z} \mid \boldsymbol{\theta}) \, p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta}^{(t)}) \, d\mathbf{Z}
\]</div>
<p><strong>Maximization step (M step):</strong> Find the parameters that maximize this quantity:</p>
<div class="math notranslate nohighlight">
\[
\boldsymbol{\theta}^{(t+1)} = \arg\max_{\boldsymbol{\theta}} Q(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})
\]</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="technical-details">
<h1>Technical Details<a class="headerlink" href="#technical-details" title="Link to this heading">#</a></h1>
<section id="why-em-works-the-monotonic-increase-guarantee">
<h2>Why EM Works: The Monotonic Increase Guarantee<a class="headerlink" href="#why-em-works-the-monotonic-increase-guarantee" title="Link to this heading">#</a></h2>
<p>The fundamental question is: why does maximizing <span class="math notranslate nohighlight">\(Q(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})\)</span> lead to an increase in the observed-data log-likelihood <span class="math notranslate nohighlight">\(\ell(\boldsymbol{\theta}; \mathbf{X}) = \log p(\mathbf{X} \mid \boldsymbol{\theta})\)</span>?</p>
<section id="the-key-decomposition">
<h3>The Key Decomposition<a class="headerlink" href="#the-key-decomposition" title="Link to this heading">#</a></h3>
<p>For any parameter values <span class="math notranslate nohighlight">\(\boldsymbol{\theta}\)</span> and <span class="math notranslate nohighlight">\(\boldsymbol{\theta}^{(t)}\)</span>, the observed-data log-likelihood can be decomposed as:</p>
<div class="math notranslate nohighlight">
\[
\ell(\boldsymbol{\theta}; \mathbf{X}) = Q(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)}) - H(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})
\]</div>
<p>where <span class="math notranslate nohighlight">\(H(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})\)</span> is the <strong>conditional entropy</strong> of the latent variables:</p>
<div class="math notranslate nohighlight">
\[
H(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)}) = \mathbb{E}_{\mathbf{Z} \sim p(\cdot \mid \mathbf{X}, \boldsymbol{\theta}^{(t)})} [\log p(\mathbf{Z} \mid \mathbf{X}, \boldsymbol{\theta})]
\]</div>
<p>The entropy term <span class="math notranslate nohighlight">\(H(\boldsymbol{\theta} \mid \boldsymbol{\theta}^{(t)})\)</span> is <strong>maximized when</strong> <span class="math notranslate nohighlight">\(\boldsymbol{\theta} = \boldsymbol{\theta}^{(t)}\)</span>.</p>
<p><strong>Each EM iteration increases (or maintains) the observed-data log-likelihood (complete proof can be found <a class="reference external" href="https://en.wikipedia.org/wiki/Expectation%E2%80%93maximization_algorithm">here</a>).</strong></p>
</section>
</section>
<section id="practical-implementation">
<h2>Practical Implementation<a class="headerlink" href="#practical-implementation" title="Link to this heading">#</a></h2>
<section id="e-step-computing-responsibilities-via-bayes-rule">
<h3>E-Step: Computing Responsibilities via Bayes’ Rule<a class="headerlink" href="#e-step-computing-responsibilities-via-bayes-rule" title="Link to this heading">#</a></h3>
<p>The E-step answers the question: “Given what I observed and my current parameter guess, how likely is each possible hidden state?”</p>
<p>For each observation <span class="math notranslate nohighlight">\(i\)</span> and each possible latent state <span class="math notranslate nohighlight">\(k\)</span>, compute:</p>
<div class="math notranslate nohighlight">
\[
w_{ik}^{(t)} = p(Z_i = k \mid X_i, \boldsymbol{\theta}^{(t)}) = \frac{p(X_i \mid Z_i = k, \boldsymbol{\theta}^{(t)}) \cdot p(Z_i = k \mid \boldsymbol{\theta}^{(t)})}{\sum_{k'} p(X_i \mid Z_i = k', \boldsymbol{\theta}^{(t)}) \cdot p(Z_i = k' \mid \boldsymbol{\theta}^{(t)})}
\]</div>
<p>These <span class="math notranslate nohighlight">\(w_{ik}^{(t)}\)</span> are posterior probability: they tell us the probability that observation <span class="math notranslate nohighlight">\(i\)</span> belongs to latent state <span class="math notranslate nohighlight">\(k\)</span>. This is just Bayes’ rule: multiply the likelihood of the observation by the prior probability of the state, then normalize.</p>
</section>
<section id="m-step-maximizing-the-q-function">
<h3>M-Step: Maximizing the Q-Function<a class="headerlink" href="#m-step-maximizing-the-q-function" title="Link to this heading">#</a></h3>
<p>The M-step finds the parameter values that best explain the data, treating the responsibilities as weights:</p>
<div class="math notranslate nohighlight">
\[
\boldsymbol{\theta}^{(t+1)} = \arg\max_{\boldsymbol{\theta}} \sum_{i=1}^n \sum_{k} w_{ik}^{(t)} \log p(X_i, Z_i = k \mid \boldsymbol{\theta})
\]</div>
<p>This is typically just weighted maximum likelihood estimation. For many models, this has simple closed-form solutions like weighted averages or weighted proportions.</p>
<p>For more detailed examples and alternative perspectives on the EM algorithm, we recommend the following references from Matthew Stephens:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://stephens999.github.io/fiveMinuteStats/em_algorithm_01.html">A prelude to the EM algorithm</a></p></li>
<li><p><a class="reference external" href="https://stephens999.github.io/fiveMinuteStats/em_variational.html">A Variational view of the EM algorithm</a></p></li>
<li><p><a class="reference external" href="https://stephens999.github.io/fiveMinuteStats/intro_to_em.html">EM in Gaussian mixture models</a></p></li>
</ul>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="related-topics">
<h1>Related Topics<a class="headerlink" href="#related-topics" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/likelihood.html">likelihood</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/maximum_likelihood_estimation.html">maximum likelihood estimation</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_rule.html">Bayes rule</a></p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="example">
<h1>Example<a class="headerlink" href="#example" title="Link to this heading">#</a></h1>
<p>Suppose we’re studying a genetic variant in a population and want to estimate its minor allele frequency (MAF). We’ve genotyped <span class="math notranslate nohighlight">\(N\)</span> individuals, but our sequencing technology isn’t perfect—it makes errors. Some individuals who are truly homozygous AA might be called as AB or BB, and vice versa.</p>
<p><strong>The problem:</strong> If we naively count alleles from the observed genotypes, our MAF estimate will be biased. How can we estimate both the true MAF and the error rate simultaneously?</p>
<p>Here we apply the EM algorithm to solve this problem. The true genotypes are our missing data—if we knew them, estimating MAF and error rate would be trivial. EM lets us handle this uncertainty.</p>
<section id="step-1-generate-simulated-data">
<h2>Step 1: Generate Simulated Data<a class="headerlink" href="#step-1-generate-simulated-data" title="Link to this heading">#</a></h2>
<p>Let’s create a dataset where we know the ground truth, so we can verify our EM algorithm works correctly.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">581</span><span class="p">)</span>
<span class="c1"># True parameters</span>
<span class="n">f_true</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.3</span><span class="w"> </span><span class="c1"># true MAF</span>
<span class="n">epsilon_true</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.05</span><span class="w"> </span><span class="c1"># true error rate</span>
<span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10000</span><span class="w"> </span><span class="c1"># sample size</span>

<span class="c1"># Step 1: Generate true genotypes according to Hardy-Weinberg equilibrium</span>
<span class="c1"># P(g=0) = (1-f)^2, P(g=1) = 2f(1-f), P(g=2) = f^2</span>
<span class="n">hw_probs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">((</span><span class="m">1</span><span class="o">-</span><span class="n">f_true</span><span class="p">)</span><span class="o">^</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="o">*</span><span class="n">f_true</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">f_true</span><span class="p">),</span><span class="w"> </span><span class="n">f_true</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>
<span class="n">true_genotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hw_probs</span><span class="p">)</span>

<span class="c1"># Step 2: Add genotyping errors</span>
<span class="n">observed_genotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">true_genotypes</span><span class="w">  </span><span class="c1"># Start with true genotypes</span>
<span class="n">errors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">epsilon_true</span><span class="w">     </span><span class="c1"># Which observations have errors?</span>

<span class="c1"># For observations with errors, randomly assign a genotype</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">errors</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">observed_genotypes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Create a data frame</span>
<span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">individual</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n</span><span class="p">,</span>
<span class="w">  </span><span class="n">true_genotype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">true_genotypes</span><span class="p">,</span>
<span class="w">  </span><span class="n">observed_genotype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">observed_genotypes</span><span class="p">,</span>
<span class="w">  </span><span class="n">error</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">errors</span>
<span class="p">)</span>

<span class="c1"># Summary</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;True MAF:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">f_true</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Naive MAF estimate:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">observed_genotypes</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">n</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Number of errors introduced:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">errors</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>True MAF: 0.3 
Naive MAF estimate: 0.31315 
Number of errors introduced: 494 
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-develop-the-em-algorithm">
<h2>Step 2: Develop the EM Algorithm<a class="headerlink" href="#step-2-develop-the-em-algorithm" title="Link to this heading">#</a></h2>
<section id="model-setup">
<h3>Model Setup<a class="headerlink" href="#model-setup" title="Link to this heading">#</a></h3>
<p><strong>Observed data:</strong> <span class="math notranslate nohighlight">\(\mathbf{d} = (d_1, \ldots, d_N)\)</span> where <span class="math notranslate nohighlight">\(d_i \in \{0, 1, 2\}\)</span></p>
<p><strong>Missing data:</strong> <span class="math notranslate nohighlight">\(\mathbf{g} = (g_1, \ldots, g_N)\)</span> where <span class="math notranslate nohighlight">\(g_i \in \{0, 1, 2\}\)</span> (true genotypes)</p>
<p><strong>Parameters:</strong> <span class="math notranslate nohighlight">\(\theta = (f, \epsilon)\)</span></p>
</section>
<section id="error-model">
<h3>Error Model<a class="headerlink" href="#error-model" title="Link to this heading">#</a></h3>
<p>The error model assumes that with probability <span class="math notranslate nohighlight">\(\epsilon\)</span>, a genotyping error occurs and the observed genotype is randomly assigned to {0, 1, 2} with equal probability <span class="math notranslate nohighlight">\(\frac{1}{3}\)</span> each. When the true genotype is <span class="math notranslate nohighlight">\(j\)</span>, observing the correct genotype <span class="math notranslate nohighlight">\(j\)</span> can happen two ways: no error (probability <span class="math notranslate nohighlight">\(1-\epsilon\)</span>) or an error that randomly lands on <span class="math notranslate nohighlight">\(j\)</span> (probability <span class="math notranslate nohighlight">\(\epsilon \times \frac{1}{3}\)</span>), giving total probability <span class="math notranslate nohighlight">\(1 - \frac{2\epsilon}{3}\)</span>. Observing an incorrect genotype <span class="math notranslate nohighlight">\(k \neq j\)</span> only happens through errors, with probability <span class="math notranslate nohighlight">\(\frac{\epsilon}{3}\)</span>.</p>
<div class="math notranslate nohighlight">
\[\begin{split}
p(d_i = k \mid g_i = j, \epsilon) = \begin{cases}
1 - \frac{2\epsilon}{3} &amp; \text{if } k = j \\
\frac{\epsilon}{3} &amp; \text{if } k \neq j
\end{cases}
\end{split}\]</div>
</section>
<section id="hardy-weinberg-model">
<h3>Hardy-Weinberg Model<a class="headerlink" href="#hardy-weinberg-model" title="Link to this heading">#</a></h3>
<div class="math notranslate nohighlight">
\[
p(g_i = 0 \mid f) = (1-f)^2, \quad p(g_i = 1 \mid f) = 2f(1-f), \quad p(g_i = 2 \mid f) = f^2
\]</div>
</section>
<section id="em-algorithm-implementation">
<h3>EM Algorithm Implementation<a class="headerlink" href="#em-algorithm-implementation" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function to compute error model probabilities</span>
<span class="n">p_obs_given_true</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># P(observed = d | true = g, epsilon)</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">d</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">g</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">2</span><span class="o">*</span><span class="n">epsilon</span><span class="o">/</span><span class="m">3</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="kr">return</span><span class="p">(</span><span class="n">epsilon</span><span class="o">/</span><span class="m">3</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">}</span>

<span class="c1"># Function to compute Hardy-Weinberg probabilities</span>
<span class="n">p_true_given_f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># P(true genotype = g | MAF = f)</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">g</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="kr">return</span><span class="p">((</span><span class="m">1</span><span class="o">-</span><span class="n">f</span><span class="p">)</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">g</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="kr">return</span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">f</span><span class="o">*</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">f</span><span class="p">))</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">g</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="kr">return</span><span class="p">(</span><span class="n">f</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># E-Step: Compute responsibilities</span>
<span class="n">e_step</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">observed</span><span class="p">)</span>
<span class="w">  </span><span class="n">w</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">  </span><span class="c1"># responsibilities for genotypes 0, 1, 2</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">d_i</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">observed</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># Compute unnormalized weights for each possible true genotype</span>
<span class="w">    </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">g</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">0</span><span class="o">:</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="m">+1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">p_obs_given_true</span><span class="p">(</span><span class="n">d_i</span><span class="p">,</span><span class="w"> </span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">p_true_given_f</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="p">)</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># Normalize</span>
<span class="w">    </span><span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">])</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># M-Step: Update parameters</span>
<span class="n">m_step</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="w"> </span><span class="n">w</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Update f (MAF)</span>
<span class="w">  </span><span class="c1"># Expected number of B alleles = sum over individuals of (1*w_i1 + 2*w_i2)</span>
<span class="w">  </span><span class="n">f_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">w</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">2</span><span class="o">*</span><span class="n">w</span><span class="p">[,</span><span class="w"> </span><span class="m">3</span><span class="p">])</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">n</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Update epsilon (error rate)</span>
<span class="w">  </span><span class="c1"># Expected number of correct calls = sum over individuals of w_{i,d_i}</span>
<span class="w">  </span><span class="n">correct_calls</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="n">w</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">observed</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">])</span>
<span class="w">  </span><span class="n">epsilon_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="m">3</span><span class="o">/</span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">correct_calls</span><span class="p">))</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">n</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f_new</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">epsilon_new</span><span class="p">))</span>
<span class="p">}</span>

<span class="c1"># Complete EM algorithm</span>
<span class="n">em_algorithm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="w"> </span><span class="n">f_init</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon_init</span><span class="p">,</span><span class="w"> </span><span class="n">max_iter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="n">tol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1e-6</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">f_init</span>
<span class="w">  </span><span class="n">epsilon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">epsilon_init</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Store history</span>
<span class="w">  </span><span class="n">history</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">    </span><span class="n">iteration</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span>
<span class="w">    </span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">,</span>
<span class="w">    </span><span class="n">epsilon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">epsilon</span><span class="p">,</span>
<span class="w">    </span><span class="n">log_likelihood</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span>
<span class="w">  </span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">iter</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">max_iter</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="c1"># E-step</span>
<span class="w">    </span><span class="n">w</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">e_step</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon</span><span class="p">)</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># M-step</span>
<span class="w">    </span><span class="n">params</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">m_step</span><span class="p">(</span><span class="n">observed</span><span class="p">,</span><span class="w"> </span><span class="n">w</span><span class="p">)</span>
<span class="w">    </span><span class="n">f_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">params</span><span class="o">$</span><span class="n">f</span>
<span class="w">    </span><span class="n">epsilon_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">params</span><span class="o">$</span><span class="n">epsilon</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># Store history</span>
<span class="w">    </span><span class="n">history</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">history</span><span class="p">,</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">      </span><span class="n">iteration</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">iter</span><span class="p">,</span>
<span class="w">      </span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f_new</span><span class="p">,</span>
<span class="w">      </span><span class="n">epsilon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">epsilon_new</span><span class="p">,</span>
<span class="w">      </span><span class="n">log_likelihood</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span>
<span class="w">    </span><span class="p">))</span>
<span class="w">    </span>
<span class="w">    </span><span class="c1"># Check convergence</span>
<span class="w">    </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">f_new</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">tol</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">epsilon_new</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">epsilon</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">tol</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Converged after&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">iter</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;iterations\n&quot;</span><span class="p">)</span>
<span class="w">      </span><span class="kr">break</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span>
<span class="w">    </span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">f_new</span>
<span class="w">    </span><span class="n">epsilon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">epsilon_new</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="kr">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span>
<span class="w">    </span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">,</span>
<span class="w">    </span><span class="n">epsilon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">epsilon</span><span class="p">,</span>
<span class="w">    </span><span class="n">history</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">history</span><span class="p">,</span>
<span class="w">    </span><span class="n">responsibilities</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">w</span>
<span class="w">  </span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-3-perform-the-em-algorithm">
<h2>Step 3: Perform the EM Algorithm<a class="headerlink" href="#step-3-perform-the-em-algorithm" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize parameters</span>
<span class="c1"># Use naive estimates as starting values</span>
<span class="n">f_init</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">observed_genotypes</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="m">2</span><span class="o">*</span><span class="n">n</span><span class="p">)</span>
<span class="n">epsilon_init</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.01</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Initial values:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;f_init =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">f_init</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;epsilon_init =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon_init</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n\n&quot;</span><span class="p">)</span>

<span class="c1"># Run EM</span>
<span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">em_algorithm</span><span class="p">(</span><span class="n">observed_genotypes</span><span class="p">,</span><span class="w"> </span><span class="n">f_init</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon_init</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Initial values:
f_init = 0.31315 
epsilon_init = 0.01 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Converged after 362 iterations
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-4-results-and-visualization">
<h2>Step 4: Results and Visualization<a class="headerlink" href="#step-4-results-and-visualization" title="Link to this heading">#</a></h2>
<p>Now let’s visualize how the EM algorithm converged to the final estimates and examine the quality of the results.</p>
<p>The convergence plots below show how the MAF and error rate estimates evolved across iterations. Notice that:</p>
<ul class="simple">
<li><p>The MAF estimate converges quickly and gets close to the true value</p></li>
<li><p>The error rate estimate also converges, but to a value lower than the true error rate</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nFinal estimates:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Estimated MAF (f) =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;(true value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">f_true</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Estimated error rate (epsilon) =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">epsilon</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;(true value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon_true</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Final estimates:
Estimated MAF (f) = 0.3027583 (true value: 0.3 )
Estimated error rate (epsilon) = 0.05268373 (true value: 0.05 )
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">history</span><span class="p">[</span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="p">],</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">iteration</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;blue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f_true</span><span class="p">,</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;dashed&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylim</span><span class="p">(</span><span class="n">f_true</span><span class="o">*</span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">f_true</span><span class="o">*</span><span class="m">1.1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">annotate</span><span class="p">(</span><span class="s">&quot;text&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f_true</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.99</span><span class="p">,</span><span class="w"> </span>
<span class="w">           </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;True value&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Convergence of MAF Estimate&quot;</span><span class="p">,</span>
<span class="w">       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Iteration&quot;</span><span class="p">,</span>
<span class="w">       </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Estimated MAF (f)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">))</span>
<span class="n">p1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/ab63499fc14649b40113045b0293c53006430860623fa4e125af8248415f9807.png"><img alt="_images/ab63499fc14649b40113045b0293c53006430860623fa4e125af8248415f9807.png" src="_images/ab63499fc14649b40113045b0293c53006430860623fa4e125af8248415f9807.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">comparison</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Parameter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;MAF (f)&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Error rate (ε)&quot;</span><span class="p">),</span>
<span class="w">  </span><span class="n">True_Value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">f_true</span><span class="p">,</span><span class="w"> </span><span class="n">epsilon_true</span><span class="p">),</span>
<span class="w">  </span><span class="n">EM_Estimate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">epsilon</span><span class="p">),</span>
<span class="w">  </span><span class="n">Absolute_Error</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">f</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">f_true</span><span class="p">),</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">epsilon</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">epsilon_true</span><span class="p">))</span>
<span class="p">)</span>
<span class="n">comparison</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 2 × 4</caption>
<thead>
	<tr><th scope=col>Parameter</th><th scope=col>True_Value</th><th scope=col>EM_Estimate</th><th scope=col>Absolute_Error</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>MAF (f)       </td><td>0.30</td><td>0.30275828</td><td>0.002758281</td></tr>
	<tr><td>Error rate (ε)</td><td>0.05</td><td>0.05268373</td><td>0.002683732</td></tr>
</tbody>
</table>
</div></div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="likelihood_ratio.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Likelihood Ratio</p>
      </div>
    </a>
    <a class="right-next"
       href="Bayesian_versus_frequentist.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bayesian versus Frequentist</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Expectation-Maximum Algorithm</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#why-em-works-the-monotonic-increase-guarantee">Why EM Works: The Monotonic Increase Guarantee</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-key-decomposition">The Key Decomposition</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#practical-implementation">Practical Implementation</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#e-step-computing-responsibilities-via-bayes-rule">E-Step: Computing Responsibilities via Bayes’ Rule</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#m-step-maximizing-the-q-function">M-Step: Maximizing the Q-Function</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-generate-simulated-data">Step 1: Generate Simulated Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-develop-the-em-algorithm">Step 2: Develop the EM Algorithm</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#model-setup">Model Setup</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#error-model">Error Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hardy-weinberg-model">Hardy-Weinberg Model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#em-algorithm-implementation">EM Algorithm Implementation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-perform-the-em-algorithm">Step 3: Perform the EM Algorithm</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-results-and-visualization">Step 4: Results and Visualization</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gao Wang and Rui Dong
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>