
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>p-value and Bayesian Hypothesis Testing &#8212; Statistical backgrounds for gene mapping</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'p_value';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Bayesian Normal Mean Model" href="Bayesian_normal_mean_model.html" />
    <link rel="prev" title="Bayes Factor" href="Bayes_factor.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="README.html">
  
  
  
  
  
  
    <p class="title logo__title">Statistical backgrounds for gene mapping</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Popular Topics and Related Notebooks</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="notations.html">Notations</a></li>
<li class="toctree-l1"><a class="reference internal" href="popular_topics.html">Popular Topics in Statistical Genetics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Basic Concepts</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genotype_coding.html">Genotype Coding</a></li>






<li class="toctree-l1"><a class="reference internal" href="minor_allele_frequency.html">Minor Allele Frequency</a></li>





<li class="toctree-l1"><a class="reference internal" href="Hardy_Weinberg_equilibrium.html">Hardy-Weinberg Equilibrium</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Variants</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium.html">Linkage Disequilibrium</a></li>






<li class="toctree-l1"><a class="reference internal" href="linkage_disequilibrium_score.html">Linkage Disequilibrium Score</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Individuals</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="genetic_relationship_matrix.html">Genetic Relationship Matrix</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Correlation Between Variables</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="factor_analysis.html">Factor Analysis</a></li>






<li class="toctree-l1"><a class="reference internal" href="principal_component_analysis.html">Principal Component Analysis (PCA)</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Fixed Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="ordinary_least_squares.html">Ordinary Least Squares</a></li>






<li class="toctree-l1"><a class="reference internal" href="odds_ratio.html">Odds Ratio</a></li>






<li class="toctree-l1"><a class="reference internal" href="marginal_joint_effects.html">Marginal and Joint Effects</a></li>






<li class="toctree-l1"><a class="reference internal" href="summary_statistics.html">Summary Statistics</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Random Effects Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="random_effect.html">Random Effect</a></li>





<li class="toctree-l1"><a class="reference internal" href="proportion_of_variance_explained.html">Proportion of Variance Explained and Heritability</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Linear Mixed Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="linear_mixed_model.html">Linear Mixed Model</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Covariates</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="confounder.html">Confounder</a></li>





<li class="toctree-l1"><a class="reference internal" href="collider.html">Collider</a></li>





<li class="toctree-l1"><a class="reference internal" href="mediator.html">Mediator</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Meta-analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="meta_analysis_fixed_effect.html">Meta-Analysis Fixed Effect</a></li>





<li class="toctree-l1"><a class="reference internal" href="meta_analysis_random_effect.html">Meta-Analysis Random Effect</a></li>





</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Likelihood</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="likelihood.html">Likelihood</a></li>






<li class="toctree-l1"><a class="reference internal" href="maximum_likelihood_estimation.html">Maximum Likelihood Estimation</a></li>






<li class="toctree-l1"><a class="reference internal" href="likelihood_ratio.html">Likelihood Ratio</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Versus Frequentist</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_frequentist.html">Bayesian and Frequentist</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bayes_rule.html">Bayes Rule</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayes_factor.html">Bayes Factor</a></li>





<li class="toctree-l1 current active"><a class="current reference internal" href="#">p-value and Bayesian Hypothesis Testing</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Bayesian Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_normal_mean_model.html">Bayesian Normal Mean Model</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_multivariate_normal_mean_model.html">Bayesian Multivariate Normal Mean Model</a></li>






</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiple Bayesian models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Bayesian_model_comparison.html">Bayesian Model Comparison</a></li>






<li class="toctree-l1"><a class="reference internal" href="Bayesian_mixture_model.html">Bayesian Mixture Model</a></li>






</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/p_value.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>p-value and Bayesian Hypothesis Testing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">p-value and Bayesian Hypothesis Testing</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value"><span class="math notranslate nohighlight">\(p\)</span>-value</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#property">Property</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#limitation-of-p-value">Limitation of <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-testing">Multiple Testing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permutation-test">Permutation test</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bayesian-hypothesis-testing">Bayesian Hypothesis Testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequentist-vs-bayesian-paradigms">Frequentist vs. Bayesian Paradigms</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#permutation-testing-for-egene-discovery">Permutation Testing for eGene Discovery</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extending-to-genome-wide-study-multiple-genes-and-fdr-control">Extending to Genome-Wide Study: Multiple Genes and FDR Control</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#supplementary">Supplementary</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Graphical Summary</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value-right-tail-probability">p-value: Right Tail Probability</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Bayesian Hypothesis Testing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Multiple Testing</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#extended-reading">Extended Reading</a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="p-value-and-bayesian-hypothesis-testing">
<h1>p-value and Bayesian Hypothesis Testing<a class="headerlink" href="#p-value-and-bayesian-hypothesis-testing" title="Link to this heading">#</a></h1>
<p>In hypothesis testing, frequentists generally use <strong>p-value</strong> to evaluate the evidence against the null hypothesis; <strong>Bayesian hypothesis testing</strong>, however, directly computes the probability of any hypothesis given observed data.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="graphical-summary">
<h1>Graphical Summary<a class="headerlink" href="#graphical-summary" title="Link to this heading">#</a></h1>
<p><img alt="Fig" src="_images/Slide25.png" /></p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="key-formula">
<h1>Key Formula<a class="headerlink" href="#key-formula" title="Link to this heading">#</a></h1>
<p>The p-value represents the probability of observing a test statistic at least as extreme as what was actually observed, assuming the null hypothesis is true.</p>
<div class="math notranslate nohighlight">
\[
p\text{-value} = P(\text{Reject } H_0 | H_0 \text{ is true})
\]</div>
<p>In Bayesian hypothesis testing,</p>
<div class="math notranslate nohighlight">
\[
P(H_i | \text{D}) = \frac{P(\text{D} | H_i) \cdot P(H_i)}{P(\text{D})}
\]</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="technical-details">
<h1>Technical Details<a class="headerlink" href="#technical-details" title="Link to this heading">#</a></h1>
<section id="p-value">
<h2><span class="math notranslate nohighlight">\(p\)</span>-value<a class="headerlink" href="#p-value" title="Link to this heading">#</a></h2>
<section id="property">
<h3>Property<a class="headerlink" href="#property" title="Link to this heading">#</a></h3>
<ol class="arabic simple">
<li><p><strong>Monotonic Relationship</strong>: Smaller p-values correspond to more extreme test statistics</p></li>
<li><p><strong>Significance Level</strong>: We reject <span class="math notranslate nohighlight">\(H_0\)</span> if <span class="math notranslate nohighlight">\(p\text{-value} \leq \alpha\)</span> (<span class="math notranslate nohighlight">\(\alpha\)</span> is the significance level)</p></li>
</ol>
</section>
<section id="limitation-of-p-value">
<h3>Limitation of <span class="math notranslate nohighlight">\(p\)</span>-value<a class="headerlink" href="#limitation-of-p-value" title="Link to this heading">#</a></h3>
<p><strong>Important</strong>: The null hypothesis <span class="math notranslate nohighlight">\(H_0\)</span> can never be “accepted” or “proven true” - it can only be:</p>
<ul class="simple">
<li><p><strong>Rejected</strong> (when <span class="math notranslate nohighlight">\(p\text{-value} \leq \alpha\)</span>): Strong evidence against <span class="math notranslate nohighlight">\(H_0\)</span></p></li>
<li><p><strong>Not rejected</strong> (when <span class="math notranslate nohighlight">\(p\text{-value} &gt; \alpha\)</span>): Insufficient evidence to reject <span class="math notranslate nohighlight">\(H_0\)</span>, doesn’t mean “accepted” or “true”. Absence of evidence is not evidence of absence.</p></li>
</ul>
</section>
<section id="multiple-testing">
<h3>Multiple Testing<a class="headerlink" href="#multiple-testing" title="Link to this heading">#</a></h3>
<p>When conducting multiple hypothesis tests, the probability of making at least one Type I error (false positive) increases dramatically. Multiple testing corrections control for this inflation.</p>
<p>In such scenarios, multiple testing correction is necessary and it has two types of correction method:</p>
<ul>
<li><p><strong>Family-Wise Error Rate (FWER) Control</strong>: control the probability of making at least one false positive across all tests.</p>
<div class="math notranslate nohighlight">
\[
  \alpha_{\text{Bonferroni}} = \frac{\alpha}{M}
  \]</div>
<p>where <span class="math notranslate nohighlight">\(M\)</span> is the total number of tests and <span class="math notranslate nohighlight">\(\alpha\)</span> is the desired overall significance level. For example, if previously we used 0.05 as the cutoff for 1 million tests, then after correction the cutoff becomes <span class="math notranslate nohighlight">\(5\times10^{-8}\)</span>.</p>
</li>
<li><p><strong>False Discovery Rate (FDR) Control</strong>: control the expected proportion of false positives among all discoveries (less stringent than FWER, thus more powerful). For example, Benjamini-Hochberg procedure controls the FDR by ranking <span class="math notranslate nohighlight">\(p\)</span>-values, finding largest <span class="math notranslate nohighlight">\(i\)</span> where <span class="math notranslate nohighlight">\(p(i) \leq (\frac{i}{M}) \times \alpha\)</span>.</p></li>
</ul>
</section>
<section id="permutation-test">
<h3>Permutation test<a class="headerlink" href="#permutation-test" title="Link to this heading">#</a></h3>
<p>Permutation test is a simple way of obtaining null distribution of a test statistics. It answers the question: “Could I have gotten this result just by random chance?” It does this by shuffling the data many times to see what results you’d get if there were truly no real effect - just randomness. The <span class="math notranslate nohighlight">\(p\)</span>-values obtained from permutation are sometimes called <em>empirical <span class="math notranslate nohighlight">\(p\)</span>-values</em>.</p>
<p><strong>Procedure</strong>:</p>
<ol class="arabic simple">
<li><p>Calculate the test statistic <span class="math notranslate nohighlight">\(T_{\text{obs}}\)</span> on the observed data</p></li>
<li><p>Randomly permute the labels/outcomes <span class="math notranslate nohighlight">\(B\)</span> times (typically <span class="math notranslate nohighlight">\(B=10,000\)</span>)</p></li>
<li><p>For each permutation <span class="math notranslate nohighlight">\(b\)</span>, calculate the test statistic <span class="math notranslate nohighlight">\(T_b\)</span></p></li>
<li><p>Compute the empirical <span class="math notranslate nohighlight">\(p\)</span>-value:
$<span class="math notranslate nohighlight">\(
p\text{-value} = \frac{1 + \sum_{b=1}^{B} \mathbb{1}(T_b \geq T_{\text{obs}})}{B + 1}
\)</span><span class="math notranslate nohighlight">\(
where \)</span>\mathbb{1}(\cdot)$ is the indicator function</p></li>
</ol>
<p><strong>Advantages</strong>:</p>
<ul class="simple">
<li><p><strong>Distribution-free</strong>: No assumptions about the underlying distribution of test statistics</p></li>
<li><p><strong>Exact for finite samples</strong>: Provides exact <span class="math notranslate nohighlight">\(p\)</span>-values rather than asymptotic approximations</p></li>
<li><p><strong>Flexible</strong>: Can be applied to any test statistic, including complex or custom statistics. Thus it is often used when the null distribution of test statistic is unknown.</p></li>
</ul>
<p><strong>Limitations</strong>:</p>
<ul class="simple">
<li><p><strong>Computationally intensive</strong>: Requires many iterations, especially problematic for large-scale studies (e.g., GWAS with millions of tests)</p></li>
<li><p><strong>Assumes exchangeability</strong>: Under <span class="math notranslate nohighlight">\(H_0\)</span>, observations must be exchangeable (permuting them doesn’t change the distribution)</p></li>
<li><p><strong>Still requires multiple testing correction</strong>: When performing many permutation tests, FWER or FDR control is still needed</p></li>
</ul>
</section>
</section>
<section id="bayesian-hypothesis-testing">
<h2>Bayesian Hypothesis Testing<a class="headerlink" href="#bayesian-hypothesis-testing" title="Link to this heading">#</a></h2>
<p>In contrast, Bayesian hypothesis testing is a lot easier to interpret, it simply means the probability of accepting the hypothesis given the observed data.</p>
</section>
<section id="frequentist-vs-bayesian-paradigms">
<h2>Frequentist vs. Bayesian Paradigms<a class="headerlink" href="#frequentist-vs-bayesian-paradigms" title="Link to this heading">#</a></h2>
<p><strong>The fundamental question</strong>: What does probability mean?</p>
<p><strong>Frequentist (<span class="math notranslate nohighlight">\(p\)</span>-value)</strong>:</p>
<ul class="simple">
<li><p>Probability = long-run frequency over repeated experiments</p></li>
<li><p>Parameters are fixed; data varies across hypothetical repetitions</p></li>
<li><p>Answers: <span class="math notranslate nohighlight">\(P(\text{Data} \mid H_0)\)</span> — “How extreme is my data if <span class="math notranslate nohighlight">\(H_0\)</span> were true?”</p></li>
<li><p><strong>Cannot say</strong>: “Probability that <span class="math notranslate nohighlight">\(H_0\)</span> is true”</p></li>
</ul>
<p><strong>Bayesian (Posterior probability)</strong>:</p>
<ul class="simple">
<li><p>Probability = degree of belief/uncertainty</p></li>
<li><p>Parameters vary; observed data is fixed</p></li>
<li><p>Answers: <span class="math notranslate nohighlight">\(P(H_0 \mid \text{Data})\)</span> — “How probable is <span class="math notranslate nohighlight">\(H_0\)</span> given my data?”</p></li>
<li><p><strong>Can say</strong>: “There’s a 95% probability the parameter is in this interval”</p></li>
</ul>
<p><strong>Multiple testing</strong>: Frequentists require explicit corrections (Bonferroni, FDR); Bayesians incorporate prior beliefs that naturally regularize.</p>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="related-topics">
<h1>Related Topics<a class="headerlink" href="#related-topics" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/ordinary_least_squares.html">ordinary least squares</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/summary_statistics.html">summary statistics</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/likelihood.html">likelihood</a></p></li>
<li><p><a class="reference external" href="https://statfungen.github.io/statgen-primer/Bayes_factor.html">Bayes factor</a></p></li>
</ul>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="example">
<h1>Example<a class="headerlink" href="#example" title="Link to this heading">#</a></h1>
<p>QTL analysis tries to answer the question of <strong>“Which genetic variants (SNPs) are associated with gene expression levels?”</strong>.</p>
<p>We’ll demonstrate this with a simple example:</p>
<ul class="simple">
<li><p>Multiple individuals with genotype data at several SNP variants</p></li>
<li><p>Gene expression measurements for a single gene</p></li>
<li><p>Testing each SNP for association with expression</p></li>
<li><p>Controlling for multiple testing using different approaches</p></li>
</ul>
<p>Let’s start with a realistic scenario: we have genotype data for 200 individuals at 15 SNPs in a region of interest, and we measure expression of a target gene.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="nf">ls</span><span class="p">())</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">83</span><span class="p">)</span>

<span class="c1"># Parameters for simulation</span>
<span class="n">n_samples</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">200</span><span class="w">      </span><span class="c1"># Number of individuals</span>
<span class="n">n_snps</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">25</span><span class="w">          </span><span class="c1"># Number of SNPs to test</span>
<span class="n">n_true_snps</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">2</span><span class="w">      </span><span class="c1"># Number of SNPs truly associated with expression</span>

<span class="c1"># Simulate genotypes with Linkage Disequilibrium (LD)</span>
<span class="c1"># Strategy: Create a causal founder haplotype and introduce recombination</span>
<span class="c1"># to create LD structure between nearby variants</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Simulating genotypes with Linkage Disequilibrium (LD)...\n&quot;</span><span class="p">)</span>

<span class="c1"># Start with a founder haplotype (allele frequencies)</span>
<span class="n">allele_freqs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n_snps</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">,</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span>

<span class="c1"># Create two haplotypes (two copies per individual)</span>
<span class="c1"># We&#39;ll use a simple LD model: nearby SNPs are correlated</span>
<span class="n">haplotype1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n_snps</span><span class="p">)</span>
<span class="n">haplotype2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n_snps</span><span class="p">)</span>

<span class="c1"># Generate first SNP independently</span>
<span class="n">haplotype1</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbinom</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">allele_freqs</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>
<span class="n">haplotype2</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbinom</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">allele_freqs</span><span class="p">[</span><span class="m">1</span><span class="p">])</span>

<span class="c1"># For subsequent SNPs, create LD with previous SNP</span>
<span class="c1"># LD correlation parameter (higher = stronger LD)</span>
<span class="n">ld_rho</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.8</span><span class="w">  </span><span class="c1"># Correlation parameter for LD (range 0-1, higher = stronger LD)</span>

<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">2</span><span class="o">:</span><span class="n">n_snps</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># Each SNP is correlated with the previous one</span>
<span class="w">  </span><span class="c1"># With probability (1-rho), SNP changes from previous; with probability rho, stays same</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Haplotype 1: LD with SNP i-1</span>
<span class="w">  </span><span class="n">haplotype1</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">haplotype1</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="m">-1</span><span class="p">]</span><span class="w">  </span><span class="c1"># Start with previous SNP allele</span>
<span class="w">  </span><span class="c1"># Randomly flip some alleles to create variation</span>
<span class="w">  </span><span class="n">flip_idx1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n_samples</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">ld_rho</span>
<span class="w">  </span><span class="n">haplotype1</span><span class="p">[</span><span class="n">flip_idx1</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">haplotype1</span><span class="p">[</span><span class="n">flip_idx1</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Haplotype 2: LD with SNP i-1 (independent from haplotype 1)</span>
<span class="w">  </span><span class="n">haplotype2</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">haplotype2</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="m">-1</span><span class="p">]</span><span class="w">  </span><span class="c1"># Start with previous SNP allele</span>
<span class="w">  </span><span class="n">flip_idx2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n_samples</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">ld_rho</span>
<span class="w">  </span><span class="n">haplotype2</span><span class="p">[</span><span class="n">flip_idx2</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">haplotype2</span><span class="p">[</span><span class="n">flip_idx2</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">]</span>
<span class="p">}</span>

<span class="c1"># Combine haplotypes to get diploid genotypes (0, 1, or 2 alternate alleles)</span>
<span class="n">genotypes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">haplotype1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">haplotype2</span>

<span class="nf">colnames</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;SNP_&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_snps</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Simulated genotype matrix with LD structure:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Dimensions:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">genotypes</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;samples x&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">ncol</span><span class="p">(</span><span class="n">genotypes</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;SNPs\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;LD correlation parameter (rho):&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">ld_rho</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;First few rows:\n&quot;</span><span class="p">)</span>
<span class="nf">head</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span>

<span class="c1"># Calculate and display LD structure (correlation matrix)</span>
<span class="n">ld_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cor</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Simulating genotypes with Linkage Disequilibrium (LD)...
Simulated genotype matrix with LD structure:
Dimensions: 200 samples x 25 SNPs
LD correlation parameter (rho): 0.8 
First few rows:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 6 × 25 of type dbl</caption>
<thead>
	<tr><th scope=col>SNP_1</th><th scope=col>SNP_2</th><th scope=col>SNP_3</th><th scope=col>SNP_4</th><th scope=col>SNP_5</th><th scope=col>SNP_6</th><th scope=col>SNP_7</th><th scope=col>SNP_8</th><th scope=col>SNP_9</th><th scope=col>SNP_10</th><th scope=col>⋯</th><th scope=col>SNP_16</th><th scope=col>SNP_17</th><th scope=col>SNP_18</th><th scope=col>SNP_19</th><th scope=col>SNP_20</th><th scope=col>SNP_21</th><th scope=col>SNP_22</th><th scope=col>SNP_23</th><th scope=col>SNP_24</th><th scope=col>SNP_25</th></tr>
</thead>
<tbody>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>1</td><td>⋯</td><td>1</td><td>1</td><td>2</td><td>2</td><td>0</td><td>1</td><td>2</td><td>1</td><td>2</td><td>1</td></tr>
	<tr><td>1</td><td>2</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>1</td><td>1</td><td>1</td><td>1</td><td>2</td><td>⋯</td><td>1</td><td>2</td><td>2</td><td>1</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>
	<tr><td>1</td><td>2</td><td>2</td><td>0</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>0</td><td>0</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>1</td><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td><td>⋯</td><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>0</td><td>1</td><td>1</td><td>1</td><td>0</td></tr>
	<tr><td>1</td><td>2</td><td>1</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>⋯</td><td>0</td><td>1</td><td>1</td><td>0</td><td>0</td><td>0</td><td>1</td><td>1</td><td>2</td><td>1</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Now we simulate expression levels. We’ll make <strong>SNP1 and SNP2 truly associated with expression</strong> (with effect sizes), and the other SNPs have no real association. The expression will also include random noise.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Simulate expression: true effects from SNP_1 and SNP_2, plus noise</span>
<span class="c1"># Expression = effect1*SNP1 + effect2*SNP2 + noise</span>

<span class="n">effect_snp1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1.5</span><span class="w">     </span><span class="c1"># Effect size for SNP_1</span>
<span class="n">effect_snp2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1.2</span><span class="w">     </span><span class="c1"># Effect size for SNP_2</span>

<span class="n">expression</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">effect_snp1</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">genotypes</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">              </span><span class="n">effect_snp2</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">genotypes</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span>
<span class="w">              </span><span class="nf">rnorm</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Simulated expression data (n =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;samples):\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;True effects: SNP_1 (effect =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">effect_snp1</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;) and SNP_2 (effect =&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">effect_snp2</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;)\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Mean expression:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">mean</span><span class="p">(</span><span class="n">expression</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;SD expression:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">sd</span><span class="p">(</span><span class="n">expression</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;First few samples:\n&quot;</span><span class="p">)</span>
<span class="nf">cbind</span><span class="p">(</span><span class="n">genotypes</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">],</span><span class="w"> </span><span class="n">expression</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">expression</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Simulated expression data (n = 200 samples):
True effects: SNP_1 (effect = 1.5 ) and SNP_2 (effect = 1.2 )
Mean expression: 1.83 
SD expression: 2.29 
First few samples:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A matrix: 5 × 6 of type dbl</caption>
<thead>
	<tr><th scope=col>SNP_1</th><th scope=col>SNP_2</th><th scope=col>SNP_3</th><th scope=col>SNP_4</th><th scope=col>SNP_5</th><th scope=col>expression</th></tr>
</thead>
<tbody>
	<tr><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td><td>1.4516029</td></tr>
	<tr><td>1</td><td>2</td><td>1</td><td>1</td><td>1</td><td>6.6169700</td></tr>
	<tr><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td><td>4.8286530</td></tr>
	<tr><td>1</td><td>2</td><td>2</td><td>0</td><td>1</td><td>4.1347712</td></tr>
	<tr><td>0</td><td>0</td><td>0</td><td>1</td><td>2</td><td>0.2250084</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>We first perform the marginal effect of each variant:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Test each SNP for association with expression</span>
<span class="c1"># Perform linear regression: expression ~ genotype for each SNP</span>

<span class="n">p_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_snps</span><span class="p">)</span>
<span class="n">effect_sizes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_snps</span><span class="p">)</span>
<span class="n">test_statistics</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_snps</span><span class="p">)</span>

<span class="c1"># Loop through each SNP and test for association</span>
<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_snps</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># Fit linear regression model: expression ~ genotype</span>
<span class="w">  </span><span class="n">model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lm</span><span class="p">(</span><span class="n">expression</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">genotypes</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="p">])</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Extract p-value from the slope coefficient test</span>
<span class="w">  </span><span class="c1"># (We want the p-value for the SNP effect, not the intercept)</span>
<span class="w">  </span><span class="n">summary_model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">summary</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
<span class="w">  </span><span class="n">p_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summary_model</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">]</span><span class="w">  </span><span class="c1"># p-value of SNP coefficient</span>
<span class="w">  </span><span class="n">effect_sizes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summary_model</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w">  </span><span class="c1"># effect size (slope)</span>
<span class="w">  </span><span class="n">test_statistics</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summary_model</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">]</span><span class="w">  </span><span class="c1"># t-statistic</span>
<span class="p">}</span>

<span class="c1"># Create a results table</span>
<span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">SNP</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">genotypes</span><span class="p">),</span>
<span class="w">  </span><span class="n">Effect</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">effect_sizes</span><span class="p">,</span>
<span class="w">  </span><span class="n">T_statistic</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_statistics</span><span class="p">,</span>
<span class="w">  </span><span class="n">P_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p_values</span><span class="p">,</span>
<span class="w">  </span><span class="n">Minus_log10_P</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">p_values</span><span class="p">)</span>
<span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Association test results for all SNPs:\n&quot;</span><span class="p">)</span>
<span class="n">results</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Association test results for all SNPs:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 25 × 5</caption>
<thead>
	<tr><th scope=col>SNP</th><th scope=col>Effect</th><th scope=col>T_statistic</th><th scope=col>P_value</th><th scope=col>Minus_log10_P</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>SNP_1 </td><td> 1.93449131</td><td> 7.6451924</td><td>8.827948e-13</td><td>12.05414024</td></tr>
	<tr><td>SNP_2 </td><td> 1.36793160</td><td> 6.3769180</td><td>1.247657e-09</td><td> 8.90390462</td></tr>
	<tr><td>SNP_3 </td><td> 0.51383262</td><td> 2.2170445</td><td>2.775781e-02</td><td> 1.55661483</td></tr>
	<tr><td>SNP_4 </td><td> 0.39898620</td><td> 1.6581493</td><td>9.887071e-02</td><td> 1.00493233</td></tr>
	<tr><td>SNP_5 </td><td> 0.08025969</td><td> 0.3528669</td><td>7.245636e-01</td><td> 0.13992350</td></tr>
	<tr><td>SNP_6 </td><td> 0.21135756</td><td> 0.8837714</td><td>3.778914e-01</td><td> 0.42263295</td></tr>
	<tr><td>SNP_7 </td><td> 0.30456995</td><td> 1.3741282</td><td>1.709554e-01</td><td> 0.76711727</td></tr>
	<tr><td>SNP_8 </td><td> 0.25950796</td><td> 1.1682952</td><td>2.440925e-01</td><td> 0.61244556</td></tr>
	<tr><td>SNP_9 </td><td>-0.03794598</td><td>-0.1692696</td><td>8.657574e-01</td><td> 0.06260378</td></tr>
	<tr><td>SNP_10</td><td> 0.07044569</td><td> 0.3073153</td><td>7.589264e-01</td><td> 0.11980036</td></tr>
	<tr><td>SNP_11</td><td> 0.39991997</td><td> 1.7391096</td><td>8.356968e-02</td><td> 1.07795129</td></tr>
	<tr><td>SNP_12</td><td> 0.06295861</td><td> 0.2760378</td><td>7.828069e-01</td><td> 0.10634536</td></tr>
	<tr><td>SNP_13</td><td> 0.42302369</td><td> 1.8880812</td><td>6.047748e-02</td><td> 1.21840634</td></tr>
	<tr><td>SNP_14</td><td> 0.24341806</td><td> 1.0512110</td><td>2.944430e-01</td><td> 0.53099881</td></tr>
	<tr><td>SNP_15</td><td> 0.12128865</td><td> 0.5156009</td><td>6.067084e-01</td><td> 0.21702000</td></tr>
	<tr><td>SNP_16</td><td> 0.07144974</td><td> 0.3084373</td><td>7.580738e-01</td><td> 0.12028849</td></tr>
	<tr><td>SNP_17</td><td>-0.10638566</td><td>-0.4891288</td><td>6.252919e-01</td><td> 0.20391722</td></tr>
	<tr><td>SNP_18</td><td>-0.29318567</td><td>-1.3328340</td><td>1.841180e-01</td><td> 0.73490385</td></tr>
	<tr><td>SNP_19</td><td>-0.22209777</td><td>-1.0453395</td><td>2.971406e-01</td><td> 0.52703800</td></tr>
	<tr><td>SNP_20</td><td>-0.10917106</td><td>-0.4787788</td><td>6.326243e-01</td><td> 0.19885416</td></tr>
	<tr><td>SNP_21</td><td> 0.13770909</td><td> 0.6009118</td><td>5.485860e-01</td><td> 0.26075527</td></tr>
	<tr><td>SNP_22</td><td>-0.07920915</td><td>-0.3418736</td><td>7.328086e-01</td><td> 0.13500947</td></tr>
	<tr><td>SNP_23</td><td>-0.20258240</td><td>-0.8236013</td><td>4.111571e-01</td><td> 0.38599225</td></tr>
	<tr><td>SNP_24</td><td>-0.08857198</td><td>-0.3854757</td><td>7.002988e-01</td><td> 0.15471663</td></tr>
	<tr><td>SNP_25</td><td>-0.22254703</td><td>-0.9874677</td><td>3.246186e-01</td><td> 0.48862665</td></tr>
</tbody>
</table>
</div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">results</span><span class="p">[</span><span class="n">results</span><span class="o">$</span><span class="n">P_value</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 3 × 5</caption>
<thead>
	<tr><th></th><th scope=col>SNP</th><th scope=col>Effect</th><th scope=col>T_statistic</th><th scope=col>P_value</th><th scope=col>Minus_log10_P</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>SNP_1</td><td>1.9344913</td><td>7.645192</td><td>8.827948e-13</td><td>12.054140</td></tr>
	<tr><th scope=row>2</th><td>SNP_2</td><td>1.3679316</td><td>6.376918</td><td>1.247657e-09</td><td> 8.903905</td></tr>
	<tr><th scope=row>3</th><td>SNP_3</td><td>0.5138326</td><td>2.217044</td><td>2.775781e-02</td><td> 1.556615</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Now let’s try to control for multiple testing by using the Bonferroni Correction, which adjusts the threshold to <span class="math notranslate nohighlight">\(\frac{\alpha}{M} = \frac{0.05}{15}\)</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply Bonferroni correction</span>
<span class="c1"># Bonferroni threshold: alpha / M</span>
<span class="n">alpha</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.05</span>
<span class="n">bonferroni_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">n_snps</span>

<span class="c1"># Identify significant SNPs under Bonferroni correction</span>
<span class="n">bonferroni_sig</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">P_value</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">bonferroni_threshold</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Bonferroni Correction (alpha = 0.05)\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;=========================================\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Adjusted significance threshold:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">bonferroni_threshold</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Number of significant SNPs:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">bonferroni_sig</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nSignificant SNPs under Bonferroni:\n&quot;</span><span class="p">)</span>
<span class="kr">if</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">bonferroni_sig</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">results</span><span class="p">[</span><span class="n">bonferroni_sig</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;(None)\n&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Bonferroni Correction (alpha = 0.05)
=========================================
Adjusted significance threshold: 0.002 
Number of significant SNPs: 2 

Significant SNPs under Bonferroni:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 2 × 5</caption>
<thead>
	<tr><th></th><th scope=col>SNP</th><th scope=col>Effect</th><th scope=col>T_statistic</th><th scope=col>P_value</th><th scope=col>Minus_log10_P</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>SNP_1</td><td>1.934491</td><td>7.645192</td><td>8.827948e-13</td><td>12.054140</td></tr>
	<tr><th scope=row>2</th><td>SNP_2</td><td>1.367932</td><td>6.376918</td><td>1.247657e-09</td><td> 8.903905</td></tr>
</tbody>
</table>
</div></div>
</div>
<p>Note that after applying Bonferroni correction, the number of significant variants decreases.</p>
<p>We can also use FDR to control the expected proportion of false positives among discoveries, allowing us to find more true associations while still controlling error.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply Benjamini-Hochberg FDR control</span>
<span class="c1"># Steps:</span>
<span class="c1"># 1. Sort p-values</span>
<span class="c1"># 2. Find largest i where p(i) &lt;= (i/M) * alpha</span>
<span class="c1"># 3. Reject all hypotheses up to index i</span>

<span class="c1"># Sort p-values and get indices</span>
<span class="n">sorted_indices</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">order</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">P_value</span><span class="p">)</span>
<span class="n">sorted_pvals</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">P_value</span><span class="p">[</span><span class="n">sorted_indices</span><span class="p">]</span>

<span class="c1"># Find the largest i where p(i) &lt;= (i/M) * alpha_FDR</span>
<span class="n">alpha_fdr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.05</span>
<span class="n">thresholds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">n_snps</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">n_snps</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">alpha_fdr</span>
<span class="n">cutoff_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">sorted_pvals</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">thresholds</span><span class="p">))</span>

<span class="c1"># If no SNP passes, cutoff_idx will be -Inf</span>
<span class="kr">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">cutoff_idx</span><span class="p">)</span><span class="w"> </span><span class="o">||</span><span class="w"> </span><span class="nf">is.infinite</span><span class="p">(</span><span class="n">cutoff_idx</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">cutoff_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="w">  </span><span class="n">fdr_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">fdr_threshold</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sorted_pvals</span><span class="p">[</span><span class="n">cutoff_idx</span><span class="p">]</span>
<span class="p">}</span>

<span class="c1"># Compute q-values (adjusted p-values for FDR)</span>
<span class="c1"># q-value for each test = min(p * M / rank, 1)</span>
<span class="n">q_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_snps</span><span class="p">)</span>
<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_snps</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">rank_i</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">sorted_indices</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">i</span><span class="p">)</span>
<span class="w">  </span><span class="n">q_values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">P_value</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">n_snps</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">rank_i</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">results</span><span class="o">$</span><span class="n">Q_value</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q_values</span>
<span class="n">fdr_sig</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">Q_value</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">alpha_fdr</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Benjamini-Hochberg FDR Control (alpha = 0.05)\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;=========================================\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;FDR threshold:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fdr_threshold</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Number of significant SNPs:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">fdr_sig</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nSignificant SNPs under FDR:\n&quot;</span><span class="p">)</span>
<span class="kr">if</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">fdr_sig</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">results</span><span class="p">[</span><span class="n">fdr_sig</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Effect&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;P_value&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Q_value&quot;</span><span class="p">)]</span>
<span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;(None)\n&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Benjamini-Hochberg FDR Control (alpha = 0.05)
=========================================
FDR threshold: 1.247657e-09 
Number of significant SNPs: 2 

Significant SNPs under FDR:
</pre></div>
</div>
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 2 × 4</caption>
<thead>
	<tr><th></th><th scope=col>SNP</th><th scope=col>Effect</th><th scope=col>P_value</th><th scope=col>Q_value</th></tr>
	<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>1</th><td>SNP_1</td><td>1.934491</td><td>8.827948e-13</td><td>2.206987e-11</td></tr>
	<tr><th scope=row>2</th><td>SNP_2</td><td>1.367932</td><td>1.247657e-09</td><td>1.559572e-08</td></tr>
</tbody>
</table>
</div></div>
</div>
<section id="permutation-testing-for-egene-discovery">
<h2>Permutation Testing for eGene Discovery<a class="headerlink" href="#permutation-testing-for-egene-discovery" title="Link to this heading">#</a></h2>
<p>Now we demonstrate permutation testing to control FDR for eGene (expression quantitative trait locus) discovery. The key idea: for each gene, we find the minimum p-value across all SNPs. We then shuffle sample labels to generate a null distribution, and compute empirical p-values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Observed test statistic: minimum p-value across all SNPs</span>
<span class="c1"># This is what we would report as the eGene p-value</span>
<span class="n">min_pval_observed</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">p_values</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Observed minimum p-value (across all SNPs):&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">min_pval_observed</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;This represents the &#39;eGene signal&#39; for this gene.\n\n&quot;</span><span class="p">)</span>

<span class="c1"># Now perform permutations to get null distribution</span>
<span class="n">n_perms</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1000</span><span class="w">  </span><span class="c1"># Number of permutations (use more for real studies, e.g., 10,000)</span>

<span class="c1"># Store minimum p-values from each permutation</span>
<span class="n">min_pvals_permuted</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_perms</span><span class="p">)</span>

<span class="nf">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">  </span><span class="c1"># For reproducibility</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Running&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n_perms</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;permutations...\n&quot;</span><span class="p">)</span>

<span class="kr">for</span><span class="p">(</span><span class="n">perm</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_perms</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="c1"># Shuffle the expression labels (break association with genotypes)</span>
<span class="w">  </span><span class="c1"># This shuffles the sample order of expression, so each SNP is tested</span>
<span class="w">  </span><span class="c1"># against a random expression ordering</span>
<span class="w">  </span><span class="n">expression_shuffled</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">expression</span><span class="p">[</span><span class="nf">sample</span><span class="p">(</span><span class="n">n_samples</span><span class="p">)]</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Test each SNP against shuffled expression</span>
<span class="w">  </span><span class="n">pvals_perm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_snps</span><span class="p">)</span>
<span class="w">  </span><span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_snps</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lm</span><span class="p">(</span><span class="n">expression_shuffled</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">genotypes</span><span class="p">[,</span><span class="w"> </span><span class="n">i</span><span class="p">])</span>
<span class="w">    </span><span class="n">summary_model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">summary</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
<span class="w">    </span><span class="n">pvals_perm</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summary_model</span><span class="o">$</span><span class="n">coefficients</span><span class="p">[</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">]</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Store minimum p-value from this permutation</span>
<span class="w">  </span><span class="n">min_pvals_permuted</span><span class="p">[</span><span class="n">perm</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">pvals_perm</span><span class="p">)</span>
<span class="p">}</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Permutation test completed.\n\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed minimum p-value (across all SNPs): 8.827948e-13 
This represents the &#39;eGene signal&#39; for this gene.

Running 1000 permutations...
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Permutation test completed.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Calculate empirical p-value</span>
<span class="c1"># p-value = (1 + # permutations with test stat &gt;= observed) / (n_perms + 1)</span>
<span class="n">empirical_pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">min_pvals_permuted</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">min_pval_observed</span><span class="p">))</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="p">(</span><span class="n">n_perms</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Empirical p-value from permutations:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">empirical_pval</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Interpretation: probability of seeing a minimum p-value this small by chance\n\n&quot;</span><span class="p">)</span>

<span class="c1"># Visualize the null distribution</span>
<span class="nf">hist</span><span class="p">(</span><span class="n">min_pvals_permuted</span><span class="p">,</span><span class="w"> </span>
<span class="w">     </span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">30</span><span class="p">,</span>
<span class="w">     </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;Null Distribution of Minimum p-value\n(from &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n_perms</span><span class="p">,</span><span class="w"> </span><span class="s">&quot; permutations)&quot;</span><span class="p">),</span>
<span class="w">     </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Minimum p-value across SNPs&quot;</span><span class="p">,</span>
<span class="w">     </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Frequency&quot;</span><span class="p">,</span>
<span class="w">     </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lightblue&quot;</span><span class="p">,</span>
<span class="w">     </span><span class="n">border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">)</span>

<span class="c1"># Add line for observed minimum p-value</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">min_pval_observed</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">lty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topright&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Observed min p-value =&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">min_pval_observed</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="p">)),</span>
<span class="w">       </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">lty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">lwd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Empirical p-value from permutations: 0.000999001 
Interpretation: probability of seeing a minimum p-value this small by chance
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/87597217d332d129b09d4d97072d42222f1e999c776ffc31461ee30c89599ab6.png"><img alt="_images/87597217d332d129b09d4d97072d42222f1e999c776ffc31461ee30c89599ab6.png" src="_images/87597217d332d129b09d4d97072d42222f1e999c776ffc31461ee30c89599ab6.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
</section>
<section id="extending-to-genome-wide-study-multiple-genes-and-fdr-control">
<h2>Extending to Genome-Wide Study: Multiple Genes and FDR Control<a class="headerlink" href="#extending-to-genome-wide-study-multiple-genes-and-fdr-control" title="Link to this heading">#</a></h2>
<p><strong>Summary of what we just did for ONE gene:</strong></p>
<ul class="simple">
<li><p>Tested 25 SNPs for association with expression</p></li>
<li><p>Used minimum p-value (min p = 8.8e-13) as the test statistic for this gene</p></li>
<li><p>Ran 1,000 permutations to generate null distribution</p></li>
<li><p>Computed empirical p-value showing this gene is a significant <strong>eGene</strong></p></li>
</ul>
<p><strong>In a real genome-wide study</strong>, we would repeat this permutation test for many genes (e.g., 100, 1000, or 10,000 genes), getting one empirical p-value per gene. Then we need to control for the multiple testing problem across genes using FDR.</p>
<p>Let’s simulate what happens in a multi-gene study:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Simulate a genome-wide eQTL study with 100 genes</span>
<span class="c1"># In reality, each gene would have its own permutation test done separately</span>

<span class="n">n_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">100</span><span class="w">  </span><span class="c1"># Number of genes tested genome-wide</span>

<span class="c1"># Our analyzed gene has this empirical p-value from permutations:</span>
<span class="n">our_gene_pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">empirical_pval</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Genome-wide FDR control for eGene discovery:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;=========================================\n\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Our gene (Gene_1) empirical p-value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="n">our_gene_pval</span><span class="p">,</span><span class="w"> </span><span class="n">scientific</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n\n&quot;</span><span class="p">)</span>

<span class="c1"># Simulate empirical p-values for 99 other genes</span>
<span class="c1"># In reality, each would have its own permutation test</span>
<span class="c1"># Most genes have no eQTL signal (p-values close to 0.5)</span>
<span class="c1"># A few have real signals (p-values close to 0)</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">72</span><span class="p">)</span>
<span class="n">other_genes_pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbeta</span><span class="p">(</span><span class="n">n_genes</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">shape1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">shape2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>

<span class="c1"># Combine: our gene + other genes</span>
<span class="n">all_genes_pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">our_gene_pval</span><span class="p">,</span><span class="w"> </span><span class="n">other_genes_pval</span><span class="p">)</span>
<span class="n">gene_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;Gene_&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_genes</span><span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Summary of empirical p-values across&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n_genes</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;genes:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- Smallest p-value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">format</span><span class="p">(</span><span class="nf">min</span><span class="p">(</span><span class="n">all_genes_pval</span><span class="p">),</span><span class="w"> </span><span class="n">scientific</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- Median p-value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">median</span><span class="p">(</span><span class="n">all_genes_pval</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- Largest p-value:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">all_genes_pval</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Genome-wide FDR control for eGene discovery:
=========================================

Our gene (Gene_1) empirical p-value: 9.99001e-04 

Summary of empirical p-values across 100 genes:
- Smallest p-value: 4.125328e-04 
- Median p-value: 0.1206 
- Largest p-value: 0.5724 
</pre></div>
</div>
</div>
</div>
<p>Now let’s apply FDR control on the gene level.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Apply FDR control (Benjamini-Hochberg) to empirical p-values</span>
<span class="c1"># This controls the expected proportion of false positive eGenes</span>
<span class="n">sorted_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">order</span><span class="p">(</span><span class="n">all_genes_pval</span><span class="p">)</span>
<span class="n">sorted_pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">all_genes_pval</span><span class="p">[</span><span class="n">sorted_idx</span><span class="p">]</span>
<span class="n">sorted_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_names</span><span class="p">[</span><span class="n">sorted_idx</span><span class="p">]</span>

<span class="c1"># Find largest i where p(i) &lt;= (i/M) * alpha</span>
<span class="n">alpha_fdr_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0.05</span>
<span class="n">thresholds_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">n_genes</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">n_genes</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">alpha_fdr_genes</span>

<span class="c1"># Find which genes pass the threshold</span>
<span class="n">sig_idx</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">sorted_pval</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">thresholds_genes</span><span class="p">)</span>

<span class="kr">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">sig_idx</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n_sig_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">sig_idx</span><span class="p">)</span>
<span class="w">  </span><span class="n">fdr_cutoff_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sorted_pval</span><span class="p">[</span><span class="n">n_sig_genes</span><span class="p">]</span>
<span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n_sig_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="w">  </span><span class="n">fdr_cutoff_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span>
<span class="p">}</span>

<span class="c1"># Compute q-values for all genes (FDR-adjusted p-values)</span>
<span class="c1"># q-value: the smallest FDR threshold at which this gene is called significant</span>
<span class="n">q_values_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">numeric</span><span class="p">(</span><span class="n">n_genes</span><span class="p">)</span>
<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_genes</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">rank_i</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">which</span><span class="p">(</span><span class="n">sorted_idx</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">i</span><span class="p">)</span>
<span class="w">  </span><span class="n">q_values_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">all_genes_pval</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">n_genes</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">rank_i</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Create results table</span>
<span class="n">genome_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">(</span>
<span class="w">  </span><span class="n">Gene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_names</span><span class="p">,</span>
<span class="w">  </span><span class="n">Empirical_Pvalue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">all_genes_pval</span><span class="p">,</span>
<span class="w">  </span><span class="n">Q_value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q_values_genes</span><span class="p">,</span>
<span class="w">  </span><span class="n">Is_eGene</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q_values_genes</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">alpha_fdr_genes</span>
<span class="p">)</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Results: Significant eGenes (q-value &lt; 0.05):\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;=========================================\n&quot;</span><span class="p">)</span>
<span class="n">egenes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genome_results</span><span class="p">[</span><span class="n">genome_results</span><span class="o">$</span><span class="n">Is_eGene</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="kr">if</span><span class="p">(</span><span class="nf">nrow</span><span class="p">(</span><span class="n">egenes</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">print</span><span class="p">(</span><span class="n">egenes</span><span class="p">)</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nTotal significant eGenes:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">egenes</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="p">}</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;(None found)\n&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nInterpretation:\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- Among&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">n_genes</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;genes tested, we found&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">egenes</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;significant eGenes\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- At FDR 0.05, we expect ~5% of these&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">egenes</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;discoveries to be false positives\n&quot;</span><span class="p">)</span>
<span class="nf">cat</span><span class="p">(</span><span class="s">&quot;- Our analyzed gene (Gene_1) is&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">any</span><span class="p">(</span><span class="n">egenes</span><span class="o">$</span><span class="n">Gene</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;Gene_1&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;significant&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;NOT significant&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Results: Significant eGenes (q-value &lt; 0.05):
=========================================
      Gene Empirical_Pvalue    Q_value Is_eGene
1   Gene_1     0.0009990010 0.04995005     TRUE
34 Gene_34     0.0004125328 0.04125328     TRUE

Total significant eGenes: 2 
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Interpretation:
- Among 100 genes tested, we found 2 significant eGenes
- At FDR 0.05, we expect ~5% of these 2 discoveries to be false positives
- Our analyzed gene (Gene_1) is significant 
</pre></div>
</div>
</div>
</div>
<p>Note that the results of gene 2-100 are artifact in this case, only to present the complete process of performing permutation test for eGene discovery.</p>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="supplementary">
<h1>Supplementary<a class="headerlink" href="#supplementary" title="Link to this heading">#</a></h1>
<section id="id1">
<h2>Graphical Summary<a class="headerlink" href="#id1" title="Link to this heading">#</a></h2>
<section id="p-value-right-tail-probability">
<h3>p-value: Right Tail Probability<a class="headerlink" href="#p-value-right-tail-probability" title="Link to this heading">#</a></h3>
<p>Visualizing the p-value as the area in the right tail of the null distribution:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="c1"># Create data for a normal distribution curve</span>
<span class="n">x_range</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">-4</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.01</span><span class="p">)</span>
<span class="n">curve_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x_range</span><span class="p">,</span><span class="w"> </span><span class="n">density</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">dnorm</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
<span class="c1"># Define the vertical line position (e.g., at x = 1.96 for a typical significance level)</span>
<span class="n">line_pos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1.96</span>
<span class="c1"># Create data for the shaded area (from line to the right)</span>
<span class="n">shade_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">curve_data</span><span class="p">[</span><span class="n">curve_data</span><span class="o">$</span><span class="n">x</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="n">line_pos</span><span class="p">,</span><span class="w"> </span><span class="p">]</span>
<span class="c1"># Create the plot</span>
<span class="n">p_value_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">curve_data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">density</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_area</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">shade_data</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#00BFC4&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">line_pos</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#00BFC4&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;dashed&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span>
<span class="w">    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expression</span><span class="p">(</span><span class="nf">italic</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="o">*</span><span class="s">&quot;-value&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Probability Density&quot;</span>
<span class="w">  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">24</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">21</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">23</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">22</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lightgray&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">),</span>
<span class="w">    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span>
<span class="w">    </span><span class="n">plot.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span>
<span class="w">  </span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="n">p_value_plot</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="s">&quot;./cartoons/p_value.png&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p_value_plot</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">units</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;in&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/d6674e431cf9d5ed505f139747b2bf4b52499f15510c4a24a39c56a85c0e83be.png" src="_images/d6674e431cf9d5ed505f139747b2bf4b52499f15510c4a24a39c56a85c0e83be.png" />
</div>
</div>
</section>
<section id="id2">
<h3>Bayesian Hypothesis Testing<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>In Bayesian hypothesis testing, we compare two hypotheses by calculating their posterior probabilities given the observed data. Below we visualize this with two overlapping normal distributions:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>

<span class="c1"># Create data for two hypothesis</span>
<span class="n">x_range</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">-4</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.01</span><span class="p">)</span>

<span class="n">h0_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span>
<span class="w">  </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x_range</span><span class="p">,</span>
<span class="w">  </span><span class="n">density</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">dnorm</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">),</span>
<span class="w">  </span><span class="n">hypothesis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Hypothesis 1&quot;</span>
<span class="p">)</span>

<span class="n">h1_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span>
<span class="w">  </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x_range</span><span class="p">,</span>
<span class="w">  </span><span class="n">density</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">dnorm</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">sd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">),</span>
<span class="w">  </span><span class="n">hypothesis</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Hypothesis 2&quot;</span>
<span class="p">)</span>

<span class="n">line_pos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">(</span><span class="m">0+5</span><span class="p">)</span><span class="o">/</span><span class="m">2</span>
<span class="n">bayes_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">h0_data</span><span class="p">,</span><span class="w"> </span><span class="n">h1_data</span><span class="p">)</span>

<span class="c1"># Create the plot</span>
<span class="n">p_Bayesian_hypothesis_testing</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">bayes_data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">density</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hypothesis</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">line_pos</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;grey&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;dashed&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span>
<span class="w">    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Bayesian Hypothesis Testing: Comparing Any Two Hypotheses&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Test Statistic&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Probability Density&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span>
<span class="w">  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">24</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">21</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="w">    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">24</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="w">    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span>
<span class="w">    </span><span class="n">plot.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span>
<span class="w">  </span><span class="p">)</span>

<span class="nf">print</span><span class="p">(</span><span class="n">p_Bayesian_hypothesis_testing</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="s">&quot;./cartoons/Bayesian_hypothesis_testing.png&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p_Bayesian_hypothesis_testing</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span>
<span class="w">       </span><span class="n">units</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;in&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/4ee914ff350af08f0ad90436d50d405efbaa0493f4678abb77096dd935c08792.png"><img alt="_images/4ee914ff350af08f0ad90436d50d405efbaa0493f4678abb77096dd935c08792.png" src="_images/4ee914ff350af08f0ad90436d50d405efbaa0493f4678abb77096dd935c08792.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
</section>
<section id="id3">
<h3>Multiple Testing<a class="headerlink" href="#id3" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>

<span class="c1"># Set seed for reproducibility</span>
<span class="nf">set.seed</span><span class="p">(</span><span class="m">71</span><span class="p">)</span>

<span class="c1"># Function to generate random sample of 10 dots (blue or red with equal probability)</span>
<span class="n">generate_sample</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">tibble</span><span class="p">(</span>
<span class="w">    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span>
<span class="w">    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">runif</span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span>
<span class="w">    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sample</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;blue&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="n">replace</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">prob</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span>
<span class="w">  </span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Function to calculate p-value using binomial test</span>
<span class="n">calc_binomial_pvalue</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">n_blue</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">data</span><span class="o">$</span><span class="n">color</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;blue&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">n_total</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">nrow</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="w">  </span><span class="c1"># Two-tailed binomial test</span>
<span class="w">  </span><span class="nf">binom.test</span><span class="p">(</span><span class="n">n_blue</span><span class="p">,</span><span class="w"> </span><span class="n">n_total</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">alternative</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;two.sided&quot;</span><span class="p">)</span><span class="o">$</span><span class="n">p.value</span>
<span class="p">}</span>

<span class="c1"># Generate 30 samples</span>
<span class="n">n_samples</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">30</span>
<span class="n">sample_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span>

<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">n_samples</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">generate_sample</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span>
<span class="w">  </span><span class="n">pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">calc_binomial_pvalue</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="w">  </span><span class="n">sample_list</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="n">pval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pval</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Create a grid of plots</span>
<span class="n">plot_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">n_samples</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sample_list</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="o">$</span><span class="n">data</span>
<span class="w">  </span><span class="n">pval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sample_list</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="o">$</span><span class="n">pval</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Count blue and red</span>
<span class="w">  </span><span class="n">n_blue</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">data</span><span class="o">$</span><span class="n">color</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;blue&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">n_red</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">data</span><span class="o">$</span><span class="n">color</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">)</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Format p-value</span>
<span class="w">  </span><span class="n">pval_text</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;p = &quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">pval</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">))</span>
<span class="w">  </span>
<span class="w">  </span><span class="c1"># Highlight significant comparisons (p &lt; 0.05)</span>
<span class="w">  </span><span class="n">border_color</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">pval</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w"> </span><span class="s">&quot;purple&quot;</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="s">&quot;gray90&quot;</span>
<span class="w">  </span><span class="n">border_size</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">pval</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="kr">else</span><span class="w"> </span><span class="m">0.5</span>
<span class="w">  </span>
<span class="w">  </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">color</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;red&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#F8766D&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;blue&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#00BFC4&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pval_text</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">theme_void</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">theme</span><span class="p">(</span>
<span class="w">      </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">margin</span><span class="p">(</span><span class="n">b</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)),</span>
<span class="w">      </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;none&quot;</span><span class="p">,</span>
<span class="w">      </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">border_color</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">border_size</span><span class="p">),</span>
<span class="w">      </span><span class="n">plot.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span>
<span class="w">      </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">),</span>
<span class="w">      </span><span class="n">plot.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">margin</span><span class="p">(</span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>
<span class="w">    </span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">    </span><span class="nf">coord_fixed</span><span class="p">(</span><span class="n">xlim</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">ylim</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
<span class="p">})</span>

<span class="c1"># Arrange in a grid (5 rows x 6 columns)</span>
<span class="n">grid_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">grid.arrange</span><span class="p">(</span><span class="n">grobs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plot_list</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">)</span>

<span class="c1"># Save</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="s">&quot;./cartoons/multiple_testing.png&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">grid_plot</span><span class="p">,</span><span class="w"> </span>
<span class="w">       </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">units</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;in&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;transparent&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<a class="reference internal image-reference" href="_images/1605c6f175715ab0f06f9088ca0979b97d6976ba9c279f80a70767f4549ce3e6.png"><img alt="_images/1605c6f175715ab0f06f9088ca0979b97d6976ba9c279f80a70767f4549ce3e6.png" src="_images/1605c6f175715ab0f06f9088ca0979b97d6976ba9c279f80a70767f4549ce3e6.png" style="width: 420px; height: 420px;" />
</a>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="extended-reading">
<h1>Extended Reading<a class="headerlink" href="#extended-reading" title="Link to this heading">#</a></h1>
<ul class="simple">
<li><p><a class="reference external" href="https://stephens999.github.io/fiveMinuteStats/BF_and_pvalue.html">Connection between Bayes factor and p-value</a></p></li>
<li><p><a class="reference external" href="https://stephens999.github.io/fiveMinuteStats/pvalue_difficult_calibrate_example.html">Example of difficulty of calibrating p values</a></p></li>
</ul>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Bayes_factor.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bayes Factor</p>
      </div>
    </a>
    <a class="right-next"
       href="Bayesian_normal_mean_model.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Bayesian Normal Mean Model</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">p-value and Bayesian Hypothesis Testing</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#graphical-summary">Graphical Summary</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#key-formula">Key Formula</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-details">Technical Details</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value"><span class="math notranslate nohighlight">\(p\)</span>-value</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#property">Property</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#limitation-of-p-value">Limitation of <span class="math notranslate nohighlight">\(p\)</span>-value</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-testing">Multiple Testing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#permutation-test">Permutation test</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#bayesian-hypothesis-testing">Bayesian Hypothesis Testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#frequentist-vs-bayesian-paradigms">Frequentist vs. Bayesian Paradigms</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#related-topics">Related Topics</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#example">Example</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#permutation-testing-for-egene-discovery">Permutation Testing for eGene Discovery</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extending-to-genome-wide-study-multiple-genes-and-fdr-control">Extending to Genome-Wide Study: Multiple Genes and FDR Control</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#supplementary">Supplementary</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Graphical Summary</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#p-value-right-tail-probability">p-value: Right Tail Probability</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Bayesian Hypothesis Testing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Multiple Testing</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#extended-reading">Extended Reading</a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gao Wang, Rui Dong
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>