{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e28d9f8f",
   "metadata": {},
   "source": [
    "# Proportion of Variance Explained and Heritability\n",
    "\n",
    "Proportion of variance explained (PVE) measures how much of the total variation in a trait (like height or disease risk) can be **attributed** to specific variables in your statistical model (e.g., genetic variants). Heritability is a specific application of this concept that measures how much of the **variation** in a trait across a population **can be explained by genetic differences**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ceb12e4",
   "metadata": {},
   "source": [
    "# Graphical Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63326357",
   "metadata": {},
   "source": [
    "![Fig](./graphical_summary/slides/Slide12.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2480c34b",
   "metadata": {},
   "source": [
    "# Key Formula\n",
    "\n",
    "Any phenotype can be modeled as the sum of genetic and environmental effects, i.e., $\\text{Phenotype}~(Y) = \\text{Genotype}~(G) + \\text{Environment}~(E)$, and under the assumption that G and E are independent from each other, the **proportion of variance explained (PVE)** by genetic effect alone (also called  broad-sense heritability $H^2$) can be derived as \n",
    "\n",
    "$$\n",
    "\\text{PVE} = H^2 = \\frac{\\text{Var}_G}{\\text{Var}_Y}\n",
    "$$\n",
    "\n",
    "where:\n",
    "- $\\text{Var}_G$ is the genetic variance component\n",
    "- $\\text{Var}_E$ is the environmental variance component"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f89a19b9",
   "metadata": {},
   "source": [
    "# Technical Details\n",
    "\n",
    "## Components of Variance\n",
    "\n",
    "Any phenotype can be modeled as the sum of genetic and environmental effects:\n",
    "\n",
    "$$\n",
    "\\text{Phenotype} (P) = \\text{Genotype} (G) + \\text{Environment} (E)\n",
    "$$\n",
    "\n",
    "The phenotypic variance in the trait can then be partitioned as:\n",
    "\n",
    "$$\\text{Var}_P = \\text{Var}_G + \\text{Var}_E + 2\\text{Cov}(G,E)$$\n",
    "\n",
    "Where:\n",
    "- $\\text{Var}_G$ is the genetic variance component\n",
    "- $\\text{Var}_E$ is the environmental variance component\n",
    "- $\\text{Cov}(G,E)$ is the covariance between genetic and environmental effects\n",
    "\n",
    "## Broad-sense Heritability\n",
    "\n",
    "In controlled experimental settings, we can design studies where $\\text{Cov}(G,E)$ is minimized and effectively set to zero. In such cases, heritability is defined as the proportion of phenotypic variance attributable to all genetic effects:\n",
    "\n",
    "$$H^2 = \\frac{\\text{Var}_G}{\\text{Var}_P}$$\n",
    "\n",
    "This represents the proportion of phenotypic variance attributable to genetic variance.\n",
    "\n",
    "## Narrow-sense Heritability\n",
    "\n",
    "**Narrow-sense heritability** ($h^2$): The proportion attributable to only **additive** genetic effects:\n",
    "\n",
    "$$h^2 = \\frac{\\text{Var}_A}{\\text{Var}_P}$$\n",
    "\n",
    "Where $\\text{Var}_A$ is the additive genetic variance, a component of $\\text{Var}_G$. \n",
    "\n",
    "Other components of $\\text{Var}_G$ includes $\\text{Var}_D$ (dominance variance) and $\\text{Var}_I$ (epistatic variance, i.e., gene-gene interaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3e9885",
   "metadata": {},
   "source": [
    "# Related Topics\n",
    "\n",
    "- [ordinary least square](https://statfungen.github.io/statgen-primer/ordinary_least_squares.html)\n",
    "- [odds ratio](https://statfungen.github.io/statgen-primer/odds_ratio.html)\n",
    "- [random effect](https://statfungen.github.io/statgen-primer/random_effect.html)\n",
    "- [LD scores](https://statfungen.github.io/statgen-primer/linkage_disequilibrium_score.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "899fd4a3-27ab-4954-a81e-e7803602a8a0",
   "metadata": {},
   "source": [
    "# Example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b096ca1f",
   "metadata": {},
   "source": [
    "When we say a trait has \"50% heritability,\" what does that actually mean? Let's explore this using the same toy example with 5 individuals and see how much of their trait variation comes from genetics versus other factors. The key question is: *How much of the differences we see in their trait values can be explained by their genetic differences?*\n",
    "\n",
    "We'll calculate this step by step, first using a simple scenario where the genetic variant has the same effect size for everyone ,then move on to the case where each person has their own unique genetic effect. This will help us understand what \"proportion of variance explained\" really means in practice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "191ec6c0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Clear the environment\n",
    "rm(list = ls())\n",
    "set.seed(11)\n",
    "# Define genotypes for 5 individuals at 3 variants\n",
    "# These represent actual alleles at each position\n",
    "# For example, Individual 1 has genotypes: CC, CT, AT\n",
    "genotypes <- c(\n",
    " \"CC\", \"CT\", \"AT\",  # Individual 1\n",
    " \"TT\", \"TT\", \"AA\",  # Individual 2\n",
    " \"CT\", \"CT\", \"AA\",  # Individual 3\n",
    " \"CC\", \"TT\", \"AA\",  # Individual 4\n",
    " \"CC\", \"CC\", \"TT\"   # Individual 5\n",
    ")\n",
    "# Reshape into a matrix\n",
    "N = 5\n",
    "M = 3\n",
    "geno_matrix <- matrix(genotypes, nrow = N, ncol = M, byrow = TRUE)\n",
    "rownames(geno_matrix) <- paste(\"Individual\", 1:N)\n",
    "colnames(geno_matrix) <- paste(\"Variant\", 1:M)\n",
    "\n",
    "alt_alleles <- c(\"T\", \"C\", \"T\")\n",
    "\n",
    "# Convert to raw genotype matrix using the additive model\n",
    "Xraw_additive <- matrix(0, nrow = N, ncol = M) # count number of non-reference alleles\n",
    "\n",
    "rownames(Xraw_additive) <- rownames(geno_matrix)\n",
    "colnames(Xraw_additive) <- colnames(geno_matrix)\n",
    "\n",
    "for (i in 1:N) {\n",
    "  for (j in 1:M) {\n",
    "    alleles <- strsplit(geno_matrix[i,j], \"\")[[1]]\n",
    "    Xraw_additive[i,j] <- sum(alleles == alt_alleles[j])\n",
    "  }\n",
    "}\n",
    "\n",
    "X <- scale(Xraw_additive, center = TRUE, scale = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd7d825b",
   "metadata": {},
   "source": [
    "## Fixed Effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "203bdc1a",
   "metadata": {},
   "source": [
    "As we discussed before, in the fixed effect model, the genetic effect $\\beta$ is a constant parameter that applies uniformly to all individuals. We then calculate the PVE of the variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d86dcc29",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.8"
      ],
      "text/latex": [
       "0.8"
      ],
      "text/markdown": [
       "0.8"
      ],
      "text/plain": [
       "[1] 0.8"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Fixed effect: single beta value for all individuals\n",
    "beta <- 0.8  # Fixed effect size\n",
    "epsilon <- rnorm(N, mean = 0, sd = 0.3)\n",
    "Y <- X[, 1] * beta + epsilon\n",
    "beta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b43bef",
   "metadata": {},
   "source": [
    "Now let's calculate the PVE using the definition as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "45c472b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.784225923945872"
      ],
      "text/latex": [
       "0.784225923945872"
      ],
      "text/markdown": [
       "0.784225923945872"
      ],
      "text/plain": [
       "[1] 0.7842259"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Calculate PVE using the definition: PVE = Var(G) / Var(Y)\n",
    "PVE <- var(X[, 1] * beta) / (var(X[, 1] * beta) + var(epsilon))\n",
    "PVE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cc514c1",
   "metadata": {},
   "source": [
    "## Random effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a9bd7d-7272-478d-85bf-e44bfd7fa712",
   "metadata": {},
   "source": [
    "Following the example in [Lecture: random effect](https://statfungen.github.io/statgen-primer/random_effect.html), we then calculate the PVE of the first variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "810af2df",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>0.438793320642639</li><li>1.48787059522956</li><li>0.0601650977954321</li><li>-0.849012867932317</li><li>2.3396930572073</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 0.438793320642639\n",
       "\\item 1.48787059522956\n",
       "\\item 0.0601650977954321\n",
       "\\item -0.849012867932317\n",
       "\\item 2.3396930572073\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 0.438793320642639\n",
       "2. 1.48787059522956\n",
       "3. 0.0601650977954321\n",
       "4. -0.849012867932317\n",
       "5. 2.3396930572073\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1]  0.4387933  1.4878706  0.0601651 -0.8490129  2.3396931"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "beta <- rnorm(N, mean = 0, sd = 1)\n",
    "beta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c8f478",
   "metadata": {},
   "source": [
    "Now let's calculate the PVE using the definition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "264c093b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.730989835353857"
      ],
      "text/latex": [
       "0.730989835353857"
      ],
      "text/markdown": [
       "0.730989835353857"
      ],
      "text/plain": [
       "[1] 0.7309898"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Calculate PVE using the definition: PVE = Var(G) / Var(Y)\n",
    "epsilon <- rnorm(N, mean = 0, sd = 0.3)\n",
    "Y <-  X[, 1] * beta + epsilon\n",
    "PVE_variant1 <- var(X[, 1] * beta) / var(Y)\n",
    "PVE_variant1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "27c3b4b7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "0.311787626831546"
      ],
      "text/latex": [
       "0.311787626831546"
      ],
      "text/markdown": [
       "0.311787626831546"
      ],
      "text/plain": [
       "[1] 0.3117876"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cov(X[, 1] * beta, epsilon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "72586d51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "2.76114631216009"
      ],
      "text/latex": [
       "2.76114631216009"
      ],
      "text/markdown": [
       "2.76114631216009"
      ],
      "text/plain": [
       "[1] 2.761146"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2.01836988811382"
      ],
      "text/latex": [
       "2.01836988811382"
      ],
      "text/markdown": [
       "2.01836988811382"
      ],
      "text/plain": [
       "[1] 2.01837"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "0.119201170383186"
      ],
      "text/latex": [
       "0.119201170383186"
      ],
      "text/markdown": [
       "0.119201170383186"
      ],
      "text/plain": [
       "[1] 0.1192012"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "2.76114631216009"
      ],
      "text/latex": [
       "2.76114631216009"
      ],
      "text/markdown": [
       "2.76114631216009"
      ],
      "text/plain": [
       "[1] 2.761146"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "var(Y)\n",
    "var(X[, 1] * beta)\n",
    "var(epsilon)\n",
    "var(X[, 1] * beta ) + var(epsilon) + 2 *cov(X[, 1] * beta, epsilon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b23e4ea5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
